// @bundle-stats/ui/lib-esm/app/app.js
var _excluded = ["jobs", "version"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { HashRouter, NavLink, Route, Switch, useLocation } from 'react-router-dom';
import { COMPONENT } from '@bundle-stats/utils';
import { METRICS_WEBPACK_ASSETS, METRICS_WEBPACK_GENERAL, METRICS_WEBPACK_MODULES, METRICS_WEBPACK_PACKAGES, URLS } from '../constants';
import { Box } from '../layout/box';
import { Container } from '../ui/container';
import { DuplicatePackagesWarning } from '../components/duplicate-packages-warning';
import { Summary } from '../components/summary';
import { BundleAssets } from '../components/bundle-assets';
import { BundleAssetsTotalsChartBars } from '../components/bundle-assets-totals-chart-bars';
import { Tabs } from '../ui/tabs';
import { Footer } from '../layout/footer';
import { Stack } from '../layout/stack';
import { BundleAssetsTotalsTable } from '../components/bundle-assets-totals-table';
import { BundleModules } from '../components/bundle-modules';
import { BundlePackages } from '../components/bundle-packages';
import { TotalSizeTypeTitle } from '../components/total-size-type-title';
import { QueryStateProvider, useComponentQueryState } from '../query-state';
import I18N from '../i18n';
import { Header } from './header';
import css from './app.module.css';

var ScrollToTop = function ScrollToTop() {
  var _useLocation = useLocation(),
      pathname = _useLocation.pathname;

  useEffect(function () {
    window.scrollTo(0, 0);
  }, [pathname]);
  return null;
};

var Layout = function Layout(_ref) {
  var jobs = _ref.jobs,
      version = _ref.version,
      props = _objectWithoutPropertiesLoose(_ref, _excluded);

  return /*#__PURE__*/React.createElement("div", {
    className: css.root
  }, /*#__PURE__*/React.createElement(Header, {
    className: css.header,
    jobs: jobs
  }), /*#__PURE__*/React.createElement("main", _extends({
    className: css.main
  }, props)), /*#__PURE__*/React.createElement(Footer, {
    version: version
  }));
};

Layout.propTypes = {
  jobs: PropTypes.arrayOf(PropTypes.object),
  version: PropTypes.string
};
Layout.defaultProps = {
  jobs: null,
  version: null
};

var AppComponent = function AppComponent(_ref2) {
  var version = _ref2.version,
      jobs = _ref2.jobs;

  var _useComponentQuerySta = useComponentQueryState(COMPONENT.BUNDLE_ASSETS),
      bundleStatsState = _useComponentQuerySta[0],
      bundleStatsSetState = _useComponentQuerySta[1];

  var _useComponentQuerySta2 = useComponentQueryState(COMPONENT.BUNDLE_PACKAGES),
      bundlePackagesState = _useComponentQuerySta2[0],
      bundlePackagesSetState = _useComponentQuerySta2[1];

  var _useComponentQuerySta3 = useComponentQueryState(COMPONENT.BUNDLE_MODULES),
      bundleModulesState = _useComponentQuerySta3[0],
      bundleModulesSetState = _useComponentQuerySta3[1];

  if (jobs.length === 0) {
    return /*#__PURE__*/React.createElement(Layout, {
      version: version
    }, /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement("div", {
      className: css.empty
    }, I18N.NO_DATA)));
  }

  var duplicatePackagesCount = jobs[0].summary.webpack.duplicatePackagesCount;
  var duplicatePackagesInsights = Boolean(duplicatePackagesCount.current || duplicatePackagesCount.baseline);
  return /*#__PURE__*/React.createElement(Layout, {
    jobs: jobs,
    version: version
  }, /*#__PURE__*/React.createElement(Container, {
    className: css.summaryContainer
  }, /*#__PURE__*/React.createElement(Summary, {
    size: "large",
    keys: METRICS_WEBPACK_GENERAL,
    data: jobs[0].summary,
    showSummaryItemDelta: jobs.length !== 1
  })), /*#__PURE__*/React.createElement(Container, {
    className: css.tabsContainer
  }, /*#__PURE__*/React.createElement(Tabs, {
    className: css.tabs
  }, /*#__PURE__*/React.createElement(NavLink, {
    exact: true,
    to: URLS.OVERVIEW,
    activeClassName: css.tabActive
  }, I18N.OVERVIEW), /*#__PURE__*/React.createElement(NavLink, {
    exact: true,
    to: URLS.ASSETS,
    activeClassName: css.tabActive
  }, I18N.ASSETS), /*#__PURE__*/React.createElement(NavLink, {
    exact: true,
    to: URLS.MODULES,
    activeClassName: css.tabActive
  }, I18N.MODULES), /*#__PURE__*/React.createElement(NavLink, {
    exact: true,
    to: URLS.PACKAGES,
    activeClassName: css.tabActive
  }, I18N.PACKAGES))), /*#__PURE__*/React.createElement("div", {
    className: css.tabsContent
  }, /*#__PURE__*/React.createElement(Switch, null, /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: URLS.ASSETS,
    render: function render(_ref3) {
      var location = _ref3.location;
      return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Stack, {
        space: "medium"
      }, /*#__PURE__*/React.createElement(Summary, {
        keys: METRICS_WEBPACK_ASSETS,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/React.createElement(Box, {
        outline: true
      }, /*#__PURE__*/React.createElement(BundleAssets, _extends({
        jobs: jobs,
        setState: bundleStatsSetState
      }, bundleStatsState, {
        key: location.pathname + "_" + location.search
      })))));
    }
  }), /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: URLS.MODULES,
    render: function render() {
      return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Stack, {
        space: "medium"
      }, /*#__PURE__*/React.createElement(Summary, {
        keys: METRICS_WEBPACK_MODULES,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/React.createElement(Box, {
        outline: true
      }, /*#__PURE__*/React.createElement(BundleModules, _extends({
        jobs: jobs,
        setState: bundleModulesSetState
      }, bundleModulesState)))));
    }
  }), /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: URLS.PACKAGES,
    render: function render(_ref4) {
      var location = _ref4.location;
      return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Stack, {
        space: "medium"
      }, /*#__PURE__*/React.createElement(Summary, {
        keys: METRICS_WEBPACK_PACKAGES,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/React.createElement(Box, {
        outline: true
      }, /*#__PURE__*/React.createElement(BundlePackages, _extends({
        jobs: jobs
      }, bundlePackagesState, {
        setState: bundlePackagesSetState,
        key: location.pathname + "_" + location.search
      })))));
    }
  }), /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: URLS.OVERVIEW,
    render: function render() {
      return /*#__PURE__*/React.createElement(Stack, {
        space: "large"
      }, duplicatePackagesInsights && /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(DuplicatePackagesWarning, {
        duplicatePackagesCount: duplicatePackagesCount
      })), /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Stack, {
        space: "small"
      }, /*#__PURE__*/React.createElement(TotalSizeTypeTitle, null), /*#__PURE__*/React.createElement(BundleAssetsTotalsChartBars, {
        jobs: jobs
      }), /*#__PURE__*/React.createElement(Box, {
        outline: true
      }, /*#__PURE__*/React.createElement(BundleAssetsTotalsTable, {
        jobs: jobs
      })))));
    }
  }))));
};

AppComponent.defaultProps = {
  jobs: [],
  version: null
};
AppComponent.propTypes = {
  jobs: PropTypes.arrayOf(PropTypes.shape({
    internalBuildNumber: PropTypes.number,
    insights: PropTypes.object,
    // eslint-disable-line react/forbid-prop-types
    summary: PropTypes.object // eslint-disable-line react/forbid-prop-types

  })),
  version: PropTypes.string
};
export var App = function App(props) {
  return /*#__PURE__*/React.createElement(HashRouter, null, /*#__PURE__*/React.createElement(ScrollToTop, null), /*#__PURE__*/React.createElement(QueryStateProvider, null, /*#__PURE__*/React.createElement(AppComponent, props)));
};
