// @bundle-stats/ui/lib/components/metrics-table-title/metrics-table-title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsTableTitle = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = _interopRequireDefault(require("../../i18n"));

var _flexStack = require("../../layout/flex-stack");

var _stack = require("../../layout/stack");

var _popover = require("../../ui/popover");

var _metricsTableTitleModule = _interopRequireDefault(require("./metrics-table-title.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var MetricsTableTitle = function MetricsTableTitle(props) {
  var className = props.className,
      title = props.title,
      info = props.info,
      popoverInfo = props.popoverInfo,
      popoverHref = props.popoverHref;
  var rootClassName = (0, _classnames["default"])(_metricsTableTitleModule["default"].root, className);
  return /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "xxxsmall",
    className: rootClassName
  }, /*#__PURE__*/_react["default"].createElement(_flexStack.FlexStack, {
    space: "xxxsmall"
  }, /*#__PURE__*/_react["default"].createElement("span", null, title), info && /*#__PURE__*/_react["default"].createElement("span", {
    className: _metricsTableTitleModule["default"].info
  }, info), (popoverInfo || popoverHref) && /*#__PURE__*/_react["default"].createElement(_popover.Popover, {
    icon: "help"
  }, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "xxxsmall"
  }, popoverInfo && /*#__PURE__*/_react["default"].createElement("p", null, popoverInfo), popoverHref && /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("a", {
    href: popoverHref,
    target: "_blank",
    rel: "noreferrer"
  }, _i18n["default"].READ_MORE))))));
};

exports.MetricsTableTitle = MetricsTableTitle;
MetricsTableTitle.propTypes = {
  className: _propTypes["default"].string,
  title: _propTypes["default"].string.isRequired,
  info: _propTypes["default"].node,
  popoverHref: _propTypes["default"].string,
  popoverInfo: _propTypes["default"].string
};
MetricsTableTitle.defaultProps = {
  className: '',
  info: null,
  popoverHref: '',
  popoverInfo: ''
};
