// @bundle-stats/plugin-webpack-validate/lib/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackSourceStruct = exports.WebpackSourceChunkStruct = exports.WebpackSourceModuleStruct = exports.WebpackSourceAssetHiddenStruct = exports.WebpackSourceAssetStruct = void 0;
const superstruct_1 = require("superstruct");
const struct = superstruct_1.superstruct({
    types: {
        notEmptyArray: (value) => (value === null || value === void 0 ? void 0 : value.length) > 0,
    },
});
exports.WebpackSourceAssetStruct = struct.interface({
    name: 'string',
    size: 'number',
});
exports.WebpackSourceAssetHiddenStruct = struct.interface({
    type: 'string',
    filteredChildren: 'number',
    size: 'number',
});
exports.WebpackSourceModuleStruct = struct.interface({
    name: 'string',
    size: 'number',
    chunks: [struct.union(['number', 'string', 'null'])],
    modules: struct.optional([struct.interface({ name: 'string', size: 'number' })]),
});
exports.WebpackSourceChunkStruct = struct.interface({
    id: struct.union(['number', 'string', 'null']),
    entry: 'boolean',
    initial: 'boolean',
    names: ['string'],
    files: ['string'],
});
exports.WebpackSourceStruct = struct.interface({
    hash: struct.optional('string'),
    builtAt: struct.optional('number'),
    assets: struct.intersection([
        [struct.union([exports.WebpackSourceAssetStruct, exports.WebpackSourceAssetHiddenStruct])],
        'notEmptyArray',
    ]),
    modules: struct.optional([exports.WebpackSourceModuleStruct]),
    chunks: struct.optional([exports.WebpackSourceChunkStruct]),
});
