// @bundle-stats/ui/lib-esm/components/metrics-table-title/metrics-table-title.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import I18N from '../../i18n';
import { FlexStack } from '../../layout/flex-stack';
import { Stack } from '../../layout/stack';
import { Popover } from '../../ui/popover';
import css from './metrics-table-title.module.css';
export var MetricsTableTitle = function MetricsTableTitle(props) {
  var className = props.className,
      title = props.title,
      info = props.info,
      popoverInfo = props.popoverInfo,
      popoverHref = props.popoverHref;
  var rootClassName = cx(css.root, className);
  return /*#__PURE__*/React.createElement(Stack, {
    space: "xxxsmall",
    className: rootClassName
  }, /*#__PURE__*/React.createElement(FlexStack, {
    space: "xxxsmall"
  }, /*#__PURE__*/React.createElement("span", null, title), info && /*#__PURE__*/React.createElement("span", {
    className: css.info
  }, info), (popoverInfo || popoverHref) && /*#__PURE__*/React.createElement(Popover, {
    icon: "help"
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "xxxsmall"
  }, popoverInfo && /*#__PURE__*/React.createElement("p", null, popoverInfo), popoverHref && /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement("a", {
    href: popoverHref,
    target: "_blank",
    rel: "noreferrer"
  }, I18N.READ_MORE))))));
};
MetricsTableTitle.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string.isRequired,
  info: PropTypes.node,
  popoverHref: PropTypes.string,
  popoverInfo: PropTypes.string
};
MetricsTableTitle.defaultProps = {
  className: '',
  info: null,
  popoverHref: '',
  popoverInfo: ''
};
