// @bundle-stats/ui/lib/layout/header/header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Header = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _headerModule = _interopRequireDefault(require("./header.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Header = function Header(_ref) {
  var className = _ref.className,
      renderLeft = _ref.renderLeft,
      renderRight = _ref.renderRight,
      render = _ref.render;
  return /*#__PURE__*/_react["default"].createElement("header", {
    className: (0, _classnames["default"])(_headerModule["default"].root, className)
  }, renderLeft({
    className: _headerModule["default"].left
  }), render({
    className: _headerModule["default"].center
  }), renderRight({
    className: _headerModule["default"].right
  }));
};

exports.Header = Header;
Header.defaultProps = {
  className: '',
  renderLeft: function renderLeft(props) {
    return /*#__PURE__*/_react["default"].createElement("div", props);
  },
  render: function render(props) {
    return /*#__PURE__*/_react["default"].createElement("div", props);
  },
  renderRight: function renderRight(props) {
    return /*#__PURE__*/_react["default"].createElement("div", props);
  }
};
Header.propTypes = {
  /* Adopted child class name */
  className: _propTypes["default"].string,

  /* Render left area */
  renderLeft: _propTypes["default"].func,

  /* Render center area */
  render: _propTypes["default"].func,

  /* Render right area */
  renderRight: _propTypes["default"].func
};
