// @bundle-stats/ui/lib-esm/components/jobs-header/jobs-header.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import I18N from '../../i18n';
import { FlexStack } from '../../layout/flex-stack';
import { JobHeader } from '../job-header';
import css from './jobs-header.module.css';
export var JobsHeader = function JobsHeader(props) {
  var className = props.className,
      jobs = props.jobs;
  var rootClassName = cx(css.root, className);
  return /*#__PURE__*/React.createElement(FlexStack, {
    className: rootClassName,
    space: "none"
  }, jobs == null ? void 0 : jobs.map(function (job, index) {
    return /*#__PURE__*/React.createElement(JobHeader, {
      key: job.internalBuildNumber || index,
      className: css.item,
      job: job,
      tag: index === 0 ? I18N.CURRENT : I18N.BASELINE
    });
  }));
};
JobsHeader.propTypes = {
  /** Adopted child classname */
  className: PropTypes.string,

  /** Jobs data */
  jobs: PropTypes.arrayOf(PropTypes.shape({
    internalBuildNumber: PropTypes.number
  }))
};
JobsHeader.defaultProps = {
  className: '',
  jobs: null
};
