// @bundle-stats/ui/lib/ui/alert/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Alert = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _box = require("../../layout/box");

var _alertModule = _interopRequireDefault(require("./alert.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Alert = function Alert(props) {
  var className = props.className,
      kind = props.kind,
      children = props.children;
  return /*#__PURE__*/_react["default"].createElement(_box.Box, {
    padding: ['xsmall', 'small'],
    outline: true,
    className: (0, _classnames["default"])(_alertModule["default"].root, className, _alertModule["default"][kind])
  }, children);
};

exports.Alert = Alert;
Alert.propTypes = {
  className: _propTypes["default"].string,
  kind: _propTypes["default"].string,
  children: _propTypes["default"].node.isRequired
};
Alert.defaultProps = {
  className: '',
  kind: 'default'
};
