// @bundle-stats/ui/lib/app/app.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.App = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRouterDom = require("react-router-dom");

var _utils = require("@bundle-stats/utils");

var _constants = require("../constants");

var _box = require("../layout/box");

var _container = require("../ui/container");

var _duplicatePackagesWarning = require("../components/duplicate-packages-warning");

var _summary = require("../components/summary");

var _bundleAssets = require("../components/bundle-assets");

var _bundleAssetsTotalsChartBars = require("../components/bundle-assets-totals-chart-bars");

var _tabs = require("../ui/tabs");

var _footer = require("../layout/footer");

var _stack = require("../layout/stack");

var _bundleAssetsTotalsTable = require("../components/bundle-assets-totals-table");

var _bundleModules = require("../components/bundle-modules");

var _bundlePackages = require("../components/bundle-packages");

var _totalSizeTypeTitle = require("../components/total-size-type-title");

var _queryState = require("../query-state");

var _i18n = _interopRequireDefault(require("../i18n"));

var _header = require("./header");

var _appModule = _interopRequireDefault(require("./app.module.css"));

var _excluded = ["jobs", "version"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ScrollToTop = function ScrollToTop() {
  var _useLocation = (0, _reactRouterDom.useLocation)(),
      pathname = _useLocation.pathname;

  (0, _react.useEffect)(function () {
    window.scrollTo(0, 0);
  }, [pathname]);
  return null;
};

var Layout = function Layout(_ref) {
  var jobs = _ref.jobs,
      version = _ref.version,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: _appModule["default"].root
  }, /*#__PURE__*/_react["default"].createElement(_header.Header, {
    className: _appModule["default"].header,
    jobs: jobs
  }), /*#__PURE__*/_react["default"].createElement("main", _extends({
    className: _appModule["default"].main
  }, props)), /*#__PURE__*/_react["default"].createElement(_footer.Footer, {
    version: version
  }));
};

Layout.propTypes = {
  jobs: _propTypes["default"].arrayOf(_propTypes["default"].object),
  version: _propTypes["default"].string
};
Layout.defaultProps = {
  jobs: null,
  version: null
};

var AppComponent = function AppComponent(_ref2) {
  var version = _ref2.version,
      jobs = _ref2.jobs;

  var _useComponentQuerySta = (0, _queryState.useComponentQueryState)(_utils.COMPONENT.BUNDLE_ASSETS),
      _useComponentQuerySta2 = _slicedToArray(_useComponentQuerySta, 2),
      bundleStatsState = _useComponentQuerySta2[0],
      bundleStatsSetState = _useComponentQuerySta2[1];

  var _useComponentQuerySta3 = (0, _queryState.useComponentQueryState)(_utils.COMPONENT.BUNDLE_PACKAGES),
      _useComponentQuerySta4 = _slicedToArray(_useComponentQuerySta3, 2),
      bundlePackagesState = _useComponentQuerySta4[0],
      bundlePackagesSetState = _useComponentQuerySta4[1];

  var _useComponentQuerySta5 = (0, _queryState.useComponentQueryState)(_utils.COMPONENT.BUNDLE_MODULES),
      _useComponentQuerySta6 = _slicedToArray(_useComponentQuerySta5, 2),
      bundleModulesState = _useComponentQuerySta6[0],
      bundleModulesSetState = _useComponentQuerySta6[1];

  if (jobs.length === 0) {
    return /*#__PURE__*/_react["default"].createElement(Layout, {
      version: version
    }, /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement("div", {
      className: _appModule["default"].empty
    }, _i18n["default"].NO_DATA)));
  }

  var duplicatePackagesCount = jobs[0].summary.webpack.duplicatePackagesCount;
  var duplicatePackagesInsights = Boolean(duplicatePackagesCount.current || duplicatePackagesCount.baseline);
  return /*#__PURE__*/_react["default"].createElement(Layout, {
    jobs: jobs,
    version: version
  }, /*#__PURE__*/_react["default"].createElement(_container.Container, {
    className: _appModule["default"].summaryContainer
  }, /*#__PURE__*/_react["default"].createElement(_summary.Summary, {
    size: "large",
    keys: _constants.METRICS_WEBPACK_GENERAL,
    data: jobs[0].summary,
    showSummaryItemDelta: jobs.length !== 1
  })), /*#__PURE__*/_react["default"].createElement(_container.Container, {
    className: _appModule["default"].tabsContainer
  }, /*#__PURE__*/_react["default"].createElement(_tabs.Tabs, {
    className: _appModule["default"].tabs
  }, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.NavLink, {
    exact: true,
    to: _constants.URLS.OVERVIEW,
    activeClassName: _appModule["default"].tabActive
  }, _i18n["default"].OVERVIEW), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.NavLink, {
    exact: true,
    to: _constants.URLS.ASSETS,
    activeClassName: _appModule["default"].tabActive
  }, _i18n["default"].ASSETS), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.NavLink, {
    exact: true,
    to: _constants.URLS.MODULES,
    activeClassName: _appModule["default"].tabActive
  }, _i18n["default"].MODULES), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.NavLink, {
    exact: true,
    to: _constants.URLS.PACKAGES,
    activeClassName: _appModule["default"].tabActive
  }, _i18n["default"].PACKAGES))), /*#__PURE__*/_react["default"].createElement("div", {
    className: _appModule["default"].tabsContent
  }, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Switch, null, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
    exact: true,
    path: _constants.URLS.ASSETS,
    render: function render(_ref3) {
      var location = _ref3.location;
      return /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
        space: "medium"
      }, /*#__PURE__*/_react["default"].createElement(_summary.Summary, {
        keys: _constants.METRICS_WEBPACK_ASSETS,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/_react["default"].createElement(_box.Box, {
        outline: true
      }, /*#__PURE__*/_react["default"].createElement(_bundleAssets.BundleAssets, _extends({
        jobs: jobs,
        setState: bundleStatsSetState
      }, bundleStatsState, {
        key: "".concat(location.pathname, "_").concat(location.search)
      })))));
    }
  }), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
    exact: true,
    path: _constants.URLS.MODULES,
    render: function render() {
      return /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
        space: "medium"
      }, /*#__PURE__*/_react["default"].createElement(_summary.Summary, {
        keys: _constants.METRICS_WEBPACK_MODULES,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/_react["default"].createElement(_box.Box, {
        outline: true
      }, /*#__PURE__*/_react["default"].createElement(_bundleModules.BundleModules, _extends({
        jobs: jobs,
        setState: bundleModulesSetState
      }, bundleModulesState)))));
    }
  }), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
    exact: true,
    path: _constants.URLS.PACKAGES,
    render: function render(_ref4) {
      var location = _ref4.location;
      return /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
        space: "medium"
      }, /*#__PURE__*/_react["default"].createElement(_summary.Summary, {
        keys: _constants.METRICS_WEBPACK_PACKAGES,
        data: jobs[0].summary,
        showSummaryItemDelta: jobs.length !== 1
      }), /*#__PURE__*/_react["default"].createElement(_box.Box, {
        outline: true
      }, /*#__PURE__*/_react["default"].createElement(_bundlePackages.BundlePackages, _extends({
        jobs: jobs
      }, bundlePackagesState, {
        setState: bundlePackagesSetState,
        key: "".concat(location.pathname, "_").concat(location.search)
      })))));
    }
  }), /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
    exact: true,
    path: _constants.URLS.OVERVIEW,
    render: function render() {
      return /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
        space: "large"
      }, duplicatePackagesInsights && /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement(_duplicatePackagesWarning.DuplicatePackagesWarning, {
        duplicatePackagesCount: duplicatePackagesCount
      })), /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
        space: "small"
      }, /*#__PURE__*/_react["default"].createElement(_totalSizeTypeTitle.TotalSizeTypeTitle, null), /*#__PURE__*/_react["default"].createElement(_bundleAssetsTotalsChartBars.BundleAssetsTotalsChartBars, {
        jobs: jobs
      }), /*#__PURE__*/_react["default"].createElement(_box.Box, {
        outline: true
      }, /*#__PURE__*/_react["default"].createElement(_bundleAssetsTotalsTable.BundleAssetsTotalsTable, {
        jobs: jobs
      })))));
    }
  }))));
};

AppComponent.defaultProps = {
  jobs: [],
  version: null
};
AppComponent.propTypes = {
  jobs: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    internalBuildNumber: _propTypes["default"].number,
    insights: _propTypes["default"].object,
    // eslint-disable-line react/forbid-prop-types
    summary: _propTypes["default"].object // eslint-disable-line react/forbid-prop-types

  })),
  version: _propTypes["default"].string
};

var App = function App(props) {
  return /*#__PURE__*/_react["default"].createElement(_reactRouterDom.HashRouter, null, /*#__PURE__*/_react["default"].createElement(ScrollToTop, null), /*#__PURE__*/_react["default"].createElement(_queryState.QueryStateProvider, null, /*#__PURE__*/_react["default"].createElement(AppComponent, props)));
};

exports.App = App;
