// @bundle-stats/ui/lib/layout/sub-header/sub-header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SubHeader = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _container = require("../../ui/container");

var _box = require("../box");

var _subHeaderModule = _interopRequireDefault(require("./sub-header.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var SIZE_SMALL = 'small';
var SIZE_DEFAULT = 'default';
var SIZES = [SIZE_SMALL, SIZE_DEFAULT];

var SubHeader = function SubHeader(props) {
  var className = props.className,
      size = props.size,
      subtitle = props.subtitle,
      title = props.title,
      icon = props.icon,
      children = props.children,
      rightSide = props.rightSide;
  var rootClassName = (0, _classnames["default"])(_subHeaderModule["default"].root, _subHeaderModule["default"][size], className);
  return /*#__PURE__*/_react["default"].createElement(_box.Box, {
    padding: "small",
    className: rootClassName
  }, /*#__PURE__*/_react["default"].createElement(_container.Container, null, /*#__PURE__*/_react["default"].createElement("div", {
    className: _subHeaderModule["default"].inner
  }, icon && /*#__PURE__*/_react["default"].createElement("div", {
    className: _subHeaderModule["default"].icon
  }, icon), /*#__PURE__*/_react["default"].createElement("div", {
    className: _subHeaderModule["default"].details
  }, subtitle && /*#__PURE__*/_react["default"].createElement("p", {
    className: _subHeaderModule["default"].subtitle
  }, subtitle), /*#__PURE__*/_react["default"].createElement("h1", {
    className: _subHeaderModule["default"].title
  }, title), children), rightSide && /*#__PURE__*/_react["default"].createElement("div", {
    className: _subHeaderModule["default"].rightSide
  }, rightSide))));
};

exports.SubHeader = SubHeader;
SubHeader.defaultProps = {
  className: '',
  size: SIZE_DEFAULT,
  subtitle: '',
  icon: null,
  children: null,
  rightSide: null
};
SubHeader.propTypes = {
  className: _propTypes["default"].string,
  size: _propTypes["default"].oneOf(SIZES),
  title: _propTypes["default"].string.isRequired,
  subtitle: _propTypes["default"].string,
  icon: _propTypes["default"].element,
  children: _propTypes["default"].element,
  rightSide: _propTypes["default"].element
};
