// @bundle-stats/ui/lib-esm/layout/sub-header/sub-header.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Container } from '../../ui/container';
import { Box } from '../box';
import css from './sub-header.module.css';
var SIZE_SMALL = 'small';
var SIZE_DEFAULT = 'default';
var SIZES = [SIZE_SMALL, SIZE_DEFAULT];
export var SubHeader = function SubHeader(props) {
  var className = props.className,
      size = props.size,
      subtitle = props.subtitle,
      title = props.title,
      icon = props.icon,
      children = props.children,
      rightSide = props.rightSide;
  var rootClassName = cx(css.root, css[size], className);
  return /*#__PURE__*/React.createElement(Box, {
    padding: "small",
    className: rootClassName
  }, /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement("div", {
    className: css.inner
  }, icon && /*#__PURE__*/React.createElement("div", {
    className: css.icon
  }, icon), /*#__PURE__*/React.createElement("div", {
    className: css.details
  }, subtitle && /*#__PURE__*/React.createElement("p", {
    className: css.subtitle
  }, subtitle), /*#__PURE__*/React.createElement("h1", {
    className: css.title
  }, title), children), rightSide && /*#__PURE__*/React.createElement("div", {
    className: css.rightSide
  }, rightSide))));
};
SubHeader.defaultProps = {
  className: '',
  size: SIZE_DEFAULT,
  subtitle: '',
  icon: null,
  children: null,
  rightSide: null
};
SubHeader.propTypes = {
  className: PropTypes.string,
  size: PropTypes.oneOf(SIZES),
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  icon: PropTypes.element,
  children: PropTypes.element,
  rightSide: PropTypes.element
};
