// @bundle-stats/ui/lib-esm/components/total-size-type-title/total-size-type-title.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import config from '../../config.json';
import I18N from '../../i18n';
import { FlexStack } from '../../layout/flex-stack';
import { Stack } from '../../layout/stack';
import { Popover } from '../../ui/popover';
import css from './total-size-type-title.module.css';
export var TotalSizeTypeTitle = function TotalSizeTypeTitle(_ref) {
  var className = _ref.className;
  return /*#__PURE__*/React.createElement(FlexStack, {
    space: "xxxsmall",
    className: cx(css.title, className)
  }, /*#__PURE__*/React.createElement("span", null, I18N.ASSET_TOTALS), /*#__PURE__*/React.createElement(Popover, {
    icon: "help"
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "xxxsmall"
  }, /*#__PURE__*/React.createElement("p", null, I18N.ASSET_TOTALS_INFO), /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement("a", {
    href: config.documentation.assets,
    target: "_blank",
    rel: "noreferrer"
  }, I18N.READ_MORE)))));
};
TotalSizeTypeTitle.propTypes = {
  className: PropTypes.string
};
TotalSizeTypeTitle.defaultProps = {
  className: ''
};
