// @bundle-stats/ui/lib/ui/empty-set/empty-set.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptySet = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _i18n = _interopRequireDefault(require("../../i18n"));

var _emptySetModule = _interopRequireDefault(require("./empty-set.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var EmptySet = function EmptySet(props) {
  var resources = props.resources,
      filtered = props.filtered,
      handleResetFilters = props.handleResetFilters,
      handleViewAll = props.handleViewAll;
  var message = filtered ? "No ".concat(resources, " found.") : "No ".concat(resources, " available.");
  var actions = [].concat(_toConsumableArray(handleResetFilters ? [/*#__PURE__*/_react["default"].createElement("button", {
    className: _emptySetModule["default"].action,
    type: "button",
    onClick: handleResetFilters
  }, _i18n["default"].RESET_FILTERS)] : []), _toConsumableArray(handleViewAll ? [/*#__PURE__*/_react["default"].createElement("button", {
    className: _emptySetModule["default"].action,
    type: "button",
    onClick: handleViewAll
  }, _i18n["default"].VIEW_ALL)] : []));
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: _emptySetModule["default"].root
  }, /*#__PURE__*/_react["default"].createElement("p", {
    className: _emptySetModule["default"].text
  }, message), filtered && actions.map(function (action, index) {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, index > 0 && ' or ', action);
  }));
};

exports.EmptySet = EmptySet;
EmptySet.propTypes = {
  resources: _propTypes["default"].string.isRequired,
  filtered: _propTypes["default"].bool.isRequired,
  handleResetFilters: _propTypes["default"].func,
  handleViewAll: _propTypes["default"].func
};
EmptySet.defaultProps = {
  handleResetFilters: null,
  handleViewAll: null
};
