// @bundle-stats/utils/lib/lighthouse/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.metrics = void 0;
const metrics_1 = require("../config/metrics");
exports.metrics = {
    score: {
        label: 'Score',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    performanceScore: {
        label: 'Performance',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    accessibilityScore: {
        label: 'Accessibility',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    bestPracticesScore: {
        label: 'Best Practices',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    seoScore: {
        label: 'SEO',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    pwaScore: {
        label: 'PWA',
        type: metrics_1.METRIC_TYPE_SCORE,
    },
    speedIndex: {
        label: 'Perceptual Speed Index',
        type: metrics_1.METRIC_TYPE_DURATION,
    },
    firstMeaningfulPaint: {
        label: 'First Meaningful Paint',
        type: metrics_1.METRIC_TYPE_DURATION,
    },
    timeToFirstByte: {
        label: 'Time To First Byte',
        type: metrics_1.METRIC_TYPE_DURATION,
    },
    firstInteractive: {
        label: 'First Interactive',
        type: metrics_1.METRIC_TYPE_DURATION,
    },
    totalByteWeight: {
        label: 'Total Weight',
        type: metrics_1.METRIC_TYPE_FILE_SIZE,
    },
    domSize: {
        label: 'DOM Size',
        type: metrics_1.METRIC_TYPE_NUMERIC,
    },
};
