// @bundle-stats/ui/lib/hocs/with-search.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withSearch = exports.useSearch = exports.generateState = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _merge2 = _interopRequireDefault(require("lodash/merge"));

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ACTION_SET_FILTERS = 'SET_FILTERS';
var ACTION_SET_SEARCH = 'SET_SEARCH';
var ACTION_RESET_DEFAULT = 'RESET_DEFAULT';
var ACTION_RESET_ALL = 'RESET_ALL';
var ACTION_SET = 'SET';

var getSearchPattern = function getSearchPattern(search) {
  var searchPattern = new RegExp(/.*/);

  if (!search || !search.trim()) {
    return searchPattern;
  }

  try {
    searchPattern = new RegExp(search, 'i');
  } catch (err) {
    // noop
    console.error(err); // eslint-disable-line no-console
  }

  return searchPattern;
};

var getSearchReducer = function getSearchReducer(_ref) {
  var defaultFilters = _ref.defaultFilters,
      allEntriesFilters = _ref.allEntriesFilters,
      setParentState = _ref.setParentState;
  return function (state, action) {
    var type = action.type,
        payload = action.payload;

    switch (type) {
      case ACTION_SET_FILTERS:
        {
          setParentState({
            filters: payload,
            search: state.search
          });
          return _objectSpread(_objectSpread({}, state), {}, {
            filters: payload
          });
        }

      case ACTION_SET_SEARCH:
        {
          setParentState({
            filters: state.filters,
            search: payload
          });
          return _objectSpread(_objectSpread({}, state), {}, {
            search: payload,
            searchPattern: getSearchPattern(payload)
          });
        }

      case ACTION_RESET_DEFAULT:
        {
          setParentState({
            filters: defaultFilters,
            search: ''
          });
          return {
            filters: defaultFilters,
            search: '',
            searchPattern: getSearchPattern()
          };
        }

      case ACTION_RESET_ALL:
        {
          setParentState({
            filters: allEntriesFilters,
            search: ''
          });
          return {
            filters: allEntriesFilters,
            search: '',
            searchPattern: getSearchPattern()
          };
        }

      case ACTION_SET:
        {
          return payload;
        }

      default:
        return state;
    }
  };
};

var generateState = function generateState(filters, search) {
  return {
    filters: filters,
    search: search,
    searchPattern: getSearchPattern(search)
  };
};

exports.generateState = generateState;

var useSearch = function useSearch(_ref2) {
  var setParentState = _ref2.setState,
      parentSearch = _ref2.search,
      parentFilters = _ref2.filters,
      defaultFilters = _ref2.defaultFilters,
      allEntriesFilters = _ref2.allEntriesFilters;
  // When we pass custom filters, set the other flags to true(allEntries)
  var initialFilters = parentFilters ? (0, _merge2["default"])({}, allEntriesFilters, parentFilters) : defaultFilters;

  var _useReducer = (0, _react.useReducer)(getSearchReducer({
    defaultFilters: defaultFilters,
    allEntriesFilters: allEntriesFilters,
    setParentState: setParentState
  }), generateState(initialFilters, parentSearch)),
      _useReducer2 = _slicedToArray(_useReducer, 2),
      _useReducer2$ = _useReducer2[0],
      search = _useReducer2$.search,
      searchPattern = _useReducer2$.searchPattern,
      filters = _useReducer2$.filters,
      dispatch = _useReducer2[1]; // Update state when the custom filters/search are changing - initial load or route updates


  (0, _react.useEffect)(function () {
    // Run a deep comparison to prevent circular setState triggering
    if (parentSearch === search || (0, _isEqual2["default"])(parentFilters, filters)) {
      return;
    }

    dispatch({
      type: ACTION_SET,
      payload: generateState(initialFilters, parentSearch)
    });
  }, [search, filters, parentFilters, parentSearch]);
  var handleUpdateSearch = (0, _react.useCallback)(function (newValue) {
    dispatch({
      type: ACTION_SET_SEARCH,
      payload: newValue
    });
  }, []);
  var handleUpdateFilters = (0, _react.useCallback)(function (newFilters) {
    dispatch({
      type: ACTION_SET_FILTERS,
      payload: newFilters
    });
  }, []);
  var handleResetFilters = (0, _react.useCallback)(function () {
    dispatch({
      type: ACTION_RESET_DEFAULT
    });
  }, []);
  var handleResetAllFilters = (0, _react.useCallback)(function () {
    dispatch({
      type: ACTION_RESET_ALL
    });
  }, []);
  return {
    search: search,
    updateSearch: handleUpdateSearch,
    searchPattern: searchPattern,
    updateFilters: handleUpdateFilters,
    resetFilters: handleResetFilters,
    resetAllFilters: handleResetAllFilters,
    filters: filters,
    hasActiveFilters: !(0, _isEqual2["default"])(allEntriesFilters, filters)
  };
};

exports.useSearch = useSearch;

var withSearch = function withSearch() {
  return function (BaseComponent) {
    var WithSearch = function WithSearch(props) {
      var allEntriesFilters = props.allEntriesFilters,
          defaultFilters = props.defaultFilters,
          filters = props.filters,
          search = props.search,
          setState = props.setState;
      var searchProps = useSearch({
        search: search,
        filters: filters,
        setState: setState,
        defaultFilters: defaultFilters,
        allEntriesFilters: allEntriesFilters
      });
      return /*#__PURE__*/_react["default"].createElement(BaseComponent, _extends({}, props, searchProps));
    };

    WithSearch.defaultProps = {
      filters: null,
      setState: _noop2["default"],
      search: ''
    };
    WithSearch.propTypes = {
      allEntriesFilters: _propTypes["default"].object.isRequired,
      // eslint-disable-line react/forbid-prop-types
      defaultFilters: _propTypes["default"].object.isRequired,
      // eslint-disable-line react/forbid-prop-types
      filters: _propTypes["default"].object,
      // eslint-disable-line react/forbid-prop-types
      search: _propTypes["default"].string,
      setState: _propTypes["default"].func
    };
    return WithSearch;
  };
};

exports.withSearch = withSearch;
