// @bundle-stats/ui/lib/components/module-info/module-info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModuleInfo = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _find2 = _interopRequireDefault(require("lodash/find"));

var _isEmpty2 = _interopRequireDefault(require("lodash/isEmpty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("@bundle-stats/utils");

var _stack = require("../../layout/stack");

var _fileName = require("../../ui/file-name");

var _componentLink = require("../component-link");

var _moduleInfoModule = _interopRequireDefault(require("./module-info.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var ModuleInfo = function ModuleInfo(props) {
  var className = props.className,
      item = props.item,
      labels = props.labels,
      chunks = props.chunks,
      chunkIds = props.chunkIds,
      CustomComponentLink = props.customComponentLink,
      onClick = props.onClick;
  return /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "small",
    className: className
  }, item.runs.map(function (run, index) {
    var TitleComponent = index !== 0 ? 'h4' : 'h3';
    var key = "module-info-".concat((run === null || run === void 0 ? void 0 : run.name) || index, "-").concat(index);
    return /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
      space: "xxxsmall",
      key: key
    }, /*#__PURE__*/_react["default"].createElement(TitleComponent, null, labels[index]), index === 0 && !(0, _isEmpty2["default"])(run === null || run === void 0 ? void 0 : run.chunkIds) && /*#__PURE__*/_react["default"].createElement("div", {
      className: _moduleInfoModule["default"].chunks
    }, /*#__PURE__*/_react["default"].createElement("strong", {
      className: _moduleInfoModule["default"].chunksTitle
    }, "Chunks:"), run.chunkIds.map(function (chunkId) {
      var chunk = (0, _find2["default"])(chunks, {
        id: chunkId
      });

      if (!chunk) {
        return null;
      }

      return /*#__PURE__*/_react["default"].createElement(CustomComponentLink, _extends({}, (0, _utils.getBundleModulesByChunk)(chunkIds, chunkId), {
        onClick: onClick,
        className: _moduleInfoModule["default"].chunksItem
      }), chunk.name);
    })), /*#__PURE__*/_react["default"].createElement(_fileName.FileName, {
      className: _moduleInfoModule["default"].fileName,
      as: "code",
      name: (run === null || run === void 0 ? void 0 : run.name) || '-'
    }));
  }));
};

exports.ModuleInfo = ModuleInfo;
ModuleInfo.propTypes = {
  className: _propTypes["default"].string,
  item: _propTypes["default"].shape({
    key: _propTypes["default"].string,
    runs: _propTypes["default"].arrayOf(_propTypes["default"].shape({
      key: _propTypes["default"].string,
      name: _propTypes["default"].string,
      chunkIds: _propTypes["default"].arrayOf(_propTypes["default"].string)
    }))
  }).isRequired,
  labels: _propTypes["default"].arrayOf(_propTypes["default"].string).isRequired,
  chunks: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    id: _propTypes["default"].string,
    name: _propTypes["default"].string
  })),
  chunkIds: _propTypes["default"].arrayOf(_propTypes["default"].string),
  customComponentLink: _propTypes["default"].elementType,
  onClick: _propTypes["default"].func
};
ModuleInfo.defaultProps = {
  className: '',
  chunks: [],
  chunkIds: [],
  customComponentLink: _componentLink.ComponentLink,
  onClick: _noop2["default"]
};
