// @bundle-stats/ui/lib/ui/logo/logo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Logo = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _logo = require("./logo.constants");

var _logoModule = _interopRequireDefault(require("./logo.module.css"));

var _excluded = ["className", "as", "kind"];

var _ICONS;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var GithubSvg = function GithubSvg(props) {
  return /*#__PURE__*/_react["default"].createElement("svg", props, /*#__PURE__*/_react["default"].createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M8 0C3.58 0 0 3.58 0 8C0 11.54 2.29 14.53 5.47 15.59C5.87 15.66 6.02 15.42 6.02 15.21C6.02 15.02 6.01 14.39 6.01 13.72C4 14.09 3.48 13.23 3.32 12.78C3.23 12.55 2.84 11.84 2.5 11.65C2.22 11.5 1.82 11.13 2.49 11.12C3.12 11.11 3.57 11.7 3.72 11.94C4.44 13.15 5.59 12.81 6.05 12.6C6.12 12.08 6.33 11.73 6.56 11.53C4.78 11.33 2.92 10.64 2.92 7.58C2.92 6.71 3.23 5.99 3.74 5.43C3.66 5.23 3.38 4.41 3.82 3.31C3.82 3.31 4.49 3.1 6.02 4.13C6.66 3.95 7.34 3.86 8.02 3.86C8.7 3.86 9.38 3.95 10.02 4.13C11.55 3.09 12.22 3.31 12.22 3.31C12.66 4.41 12.38 5.23 12.3 5.43C12.81 5.99 13.12 6.7 13.12 7.58C13.12 10.65 11.25 11.33 9.47 11.53C9.76 11.78 10.01 12.26 10.01 13.01C10.01 14.08 10 14.94 10 15.21C10 15.42 10.15 15.67 10.55 15.59C13.71 14.53 16 11.53 16 8C16 3.58 12.42 0 8 0Z",
    transform: "scale(64)"
  }));
};

GithubSvg.defaultProps = {
  viewBox: "0 0 1024 1024",
  fill: "currentColor",
  xmlns: "http://www.w3.org/2000/svg"
};

var LogoSvg = function LogoSvg(props) {
  return /*#__PURE__*/_react["default"].createElement("svg", props, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M256 42.67L69.33 149.33v213.34L256 469.33l186.67-106.66V149.33L256 42.67z",
    fill: "#fff"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M256 59.12l172.27 98.44v196.88L256 452.88 83.73 354.44V157.56L256 59.12M256 0L32 128v256l224 128 224-128V128L256 0z",
    fill: "#9da1a6"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M256 128l-112 64v128l112 64 112-64V192l-112-64z",
    fill: "#8b8f95"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M144 192v128l112 64V249.33z",
    fill: "#6f747b"
  }));
};

LogoSvg.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 512"
};

var LogotypeSvg = function LogotypeSvg(props) {
  return /*#__PURE__*/_react["default"].createElement("svg", props, /*#__PURE__*/_react["default"].createElement("path", {
    className: "a",
    d: "M12.68 53.56h-.33l-1.17 4.51H0V0h14.35v13.52l-.5 6h.33a15.46 15.46 0 0110.43-4.17c10.1 0 16.61 8.34 16.61 21.11 0 14.43-8.68 22.61-17.77 22.61-3.76 0-7.51-1.83-10.77-5.51zm13.85-16.85c0-6.84-2.08-9.68-6.17-9.68-2.25 0-4 .84-6 3v15.36a8.47 8.47 0 005.67 2c3.5 0 6.5-2.67 6.5-10.68zM48.57 42.05v-25.7h14.35v23.86c0 4.68 1 6.68 4.76 6.68 2.42 0 4.08-.67 6.25-3.67V16.35h14.35v41.72H76.6l-1-5.51h-.33c-3.51 4-7.68 6.51-13.52 6.51-9.34 0-13.18-6.67-13.18-17.02zM97.4 16.35h11.68l1 5h.33c3.34-3.17 7.85-6 13.94-6 9.26 0 13.26 6.68 13.26 17v25.7h-14.35V34.21c0-4.67-1.16-6.68-4.84-6.68-2.67 0-4.5 1.17-6.67 3.17v27.37H97.4zM144.8 37.21c0-13.68 8.68-21.86 17.44-21.86a12.27 12.27 0 019.76 4.17h.33l-.66-6V0H186v58.07h-11.66l-1-3.84H173c-2.92 2.84-7 4.84-10.76 4.84-10.6 0-17.44-8.34-17.44-21.86zm26.87 7.18V29a7.8 7.8 0 00-5.51-2c-3.5 0-6.67 2.67-6.67 10 0 7.51 2.41 10.35 6.5 10.35a6.15 6.15 0 005.68-2.96zM204.05 40.88v-29.7h-12.51V0h26.86v41.38c0 4.84 3.34 6.18 6.26 6.18a18.15 18.15 0 006-1.17l3.08 10.51a39.1 39.1 0 01-12.93 2.17c-11.09 0-16.76-7.34-16.76-18.19zM238 37.21c0-13.68 10.68-21.86 21.86-21.86 13.18 0 19.85 9 19.85 20.19a37.41 37.41 0 01-.58 6h-30.25V32.2h20.36l-2 2.51c0-5.51-2.17-8.51-6.84-8.51-5.51 0-8.51 3.5-8.51 11 0 7.84 3.84 11 11.51 11a23.26 23.26 0 0010-2.5l4.68 8.51a33.13 33.13 0 01-16.69 4.84c-13.17.02-23.39-7.99-23.39-21.84zM284.18 51.31l8.18-9.84c4.08 3.08 8.92 5.25 13.85 5.25 4.33 0 6-1.42 6-3.17 0-3.17-2.83-3.84-7.84-5.92l-6.76-2.84c-6.26-2.33-11.43-7.26-11.43-15.1 0-9.26 8.34-16.85 20.69-16.85a27.27 27.27 0 0118.53 7.25l-7.18 9c-3.92-2.59-7-3.93-12.18-3.93-3.17 0-5.34 1.17-5.34 3.68s3.09 3.42 8.51 5.42l6.26 2.59c7.26 2.67 11.43 7.34 11.43 14.85 0 9.17-7.68 17.35-21.7 17.35a32.65 32.65 0 01-21.02-7.74zM337.62 40.38V27.53h-10V16.85l10.68-.5 1.84-12.51H352v12.51h16.52v11.18H352v12.6c0 5.93 2.76 7.76 8.1 7.76a23.71 23.71 0 007.59-1.33l2.5 10.34a50.19 50.19 0 01-13.85 2.17c-13.71 0-18.72-7.51-18.72-18.69zM376.09 46.22c0-9.34 7.34-14.18 26.37-15.18-.51-3-3.34-4.34-8-4.34-3.51 0-7 1.17-12 3.5l-5-9.34a40.45 40.45 0 0119.86-5.51c11.85 0 19.53 5.84 19.53 19.52v23.2h-11.72l-1-4.34h-.33c-4.09 3-9 5.34-14.18 5.34-8.69 0-13.53-6.01-13.53-12.85zm26.37-1.33v-5.34c-10.18.33-12.69 2.67-12.69 5.5 0 1.92 1.5 3 4.51 3 3.34.01 5.92-1.49 8.18-3.16zM431.6 40.38V27.53h-10V16.85l10.68-.5 1.82-12.51H446v12.51h16.52v11.18H446v12.6c0 5.93 2.75 7.76 8.09 7.76a23.77 23.77 0 007.6-1.33l2.5 10.34a50.26 50.26 0 01-13.85 2.17c-13.73 0-18.74-7.51-18.74-18.69zM470.28 52.73l6-8.84a32.83 32.83 0 0016.35 4.67c3.68 0 5.18-1 5.18-2.5 0-2-2.17-2.34-9.18-3.84-11.5-2.51-15.63-6.84-15.63-13.35 0-8.18 7.76-13.52 19.86-13.52 7.84 0 14.1 3.09 18.52 5.84L505 29.7a27.79 27.79 0 00-13.51-3.84c-2.67 0-4.34.84-4.34 2.34 0 1.84 2.17 2 8.67 3.5C508.33 34.54 512 38.3 512 45.22c0 7.84-7.43 13.85-21.36 13.85a39.71 39.71 0 01-20.36-6.34z"
  }));
};

LogotypeSvg.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 512 59.07",
  fill: "currentColor"
};

var WebpackSvg = function WebpackSvg(props) {
  return /*#__PURE__*/_react["default"].createElement("svg", props, /*#__PURE__*/_react["default"].createElement("path", {
    fill: "#FFF",
    d: "M300 .1L565 150v299.9L300 599.8 35 449.9V150z"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    fill: "#8ED6FB",
    d: "M517.7 439.5L308.8 557.8v-92L439 394.1l78.7 45.4zm14.3-12.9V179.4l-76.4 44.1v159l76.4 44.1zM81.5 439.5l208.9 118.2v-92l-130.2-71.6-78.7 45.4zm-14.3-12.9V179.4l76.4 44.1v159l-76.4 44.1zm8.9-263.2L290.4 42.2v89l-137.3 75.5-1.1.6-75.9-43.9zm446.9 0L308.8 42.2v89L446 206.8l1.1.6 75.9-44z"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    fill: "#1C78C0",
    d: "M290.4 444.8L162 374.1V234.2l128.4 74.1v136.5zm18.4 0l128.4-70.6v-140l-128.4 74.1v136.5zM299.6 303zm-129-85l129-70.9L428.5 218l-128.9 74.4-129-74.4z"
  }));
};

WebpackSvg.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 600 600"
};
var ICONS = (_ICONS = {}, _defineProperty(_ICONS, _logo.KIND_GITHUB, GithubSvg), _defineProperty(_ICONS, _logo.KIND_LOGOTYPE, LogotypeSvg), _defineProperty(_ICONS, _logo.KIND_LOGO, LogoSvg), _defineProperty(_ICONS, _logo.KIND_WEBPACK, WebpackSvg), _ICONS);

var Logo = function Logo(_ref) {
  var className = _ref.className,
      Component = _ref.as,
      kind = _ref.kind,
      props = _objectWithoutProperties(_ref, _excluded);

  var rootClassName = (0, _classnames["default"])(_logoModule["default"].root, _logoModule["default"][kind], className);
  var Svg = ICONS[kind] || LogoSvg;
  return /*#__PURE__*/_react["default"].createElement(Component, _extends({
    className: rootClassName
  }, props), /*#__PURE__*/_react["default"].createElement(Svg, {
    className: _logoModule["default"].svg
  }));
};

exports.Logo = Logo;
Logo.defaultProps = {
  as: 'span',
  className: '',
  kind: _logo.KIND_DEFAULT
};
Logo.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string,

  /** Rendered component */
  as: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].elementType]),

  /** Kind modifier */
  kind: _propTypes["default"].oneOf(_logo.KINDS)
};
