// @bundle-stats/ui/lib-esm/components/job-name/job-name.js
var _excluded = ["as", "title", "internalBuildNumber", "children"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import { Tooltip } from '../../ui/tooltip';
import css from './job-name.module.css';
export var JobName = function JobName(_ref) {
  var Component = _ref.as,
      title = _ref.title,
      internalBuildNumber = _ref.internalBuildNumber,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, _excluded);

  return /*#__PURE__*/React.createElement(Tooltip, _extends({
    className: css.root,
    as: Component,
    title: title
  }, restProps), children || "Job #" + internalBuildNumber);
};
JobName.defaultProps = {
  as: 'span',
  title: '',
  internalBuildNumber: null,
  children: null
};
JobName.propTypes = {
  /** Rendered component */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /** Tooltip title */
  title: PropTypes.string,

  /** Job internal build number */
  internalBuildNumber: PropTypes.number,

  /** Render function */
  children: PropTypes.node
};
