// @bundle-stats/ui/lib/layout/footer/footer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Footer = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _config = _interopRequireDefault(require("../../config.json"));

var _container = require("../../ui/container");

var _box = require("../box");

var _stack = require("../stack");

var _footerModule = _interopRequireDefault(require("./footer.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/* eslint-disable jsx-a11y/anchor-has-content */
var ExternalLink = function ExternalLink(props) {
  return /*#__PURE__*/_react["default"].createElement("a", _extends({
    target: "_blank",
    rel: "noopener noreferrer"
  }, props));
};
/* eslint-enable jsx-a11y/anchor-has-content */


ExternalLink.propTypes = {
  /** Link to value */
  to: _propTypes["default"].string.isRequired
};

var Footer = function Footer(_ref) {
  var className = _ref.className,
      version = _ref.version;
  return /*#__PURE__*/_react["default"].createElement(_container.Container, {
    as: "footer",
    className: (0, _classnames["default"])(_footerModule["default"].root, className)
  }, /*#__PURE__*/_react["default"].createElement(_box.Box, {
    padding: ['medium', 'none'],
    className: _footerModule["default"].inner
  }, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "xsmall"
  }, /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("nav", {
    className: _footerModule["default"].nav
  }, /*#__PURE__*/_react["default"].createElement(ExternalLink, {
    href: _config["default"].githubUrl,
    className: (0, _classnames["default"])(_footerModule["default"].navItem, _footerModule["default"].navLink),
    title: "In-depth bundle analyzer for webpack(bundle size, assets, modules, packages)"
  }, "BundleStats ", version), /*#__PURE__*/_react["default"].createElement("span", {
    className: _footerModule["default"].navItem
  }, /*#__PURE__*/_react["default"].createElement(ExternalLink, {
    href: "".concat(_config["default"].githubUrl, "/issues"),
    className: _footerModule["default"].navLink,
    title: "Navigate to GitHub and open issue"
  }, "Report issue")), /*#__PURE__*/_react["default"].createElement("span", {
    className: _footerModule["default"].navItem
  }, /*#__PURE__*/_react["default"].createElement("span", null, "Made by "), /*#__PURE__*/_react["default"].createElement(ExternalLink, {
    className: _footerModule["default"].navLink,
    href: _config["default"].siteUrl,
    title: "Go to RelativeCI - Specialized insights for web bundles"
  }, "RelativeCI")))))));
};

exports.Footer = Footer;
Footer.defaultProps = {
  className: '',
  version: null
};
Footer.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string,

  /** Content */
  version: _propTypes["default"].string
};
