// use-query-params/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryParamProvider = exports.QueryParams = exports.withQueryParamsMapped = exports.withQueryParams = exports.useQueryParams = exports.useQueryParam = void 0;
__exportStar(require("serialize-query-params"), exports);
__exportStar(require("./types"), exports);
var useQueryParam_1 = require("./useQueryParam");
Object.defineProperty(exports, "useQueryParam", { enumerable: true, get: function () { return useQueryParam_1.useQueryParam; } });
var useQueryParams_1 = require("./useQueryParams");
Object.defineProperty(exports, "useQueryParams", { enumerable: true, get: function () { return useQueryParams_1.useQueryParams; } });
var withQueryParams_1 = require("./withQueryParams");
Object.defineProperty(exports, "withQueryParams", { enumerable: true, get: function () { return withQueryParams_1.withQueryParams; } });
Object.defineProperty(exports, "withQueryParamsMapped", { enumerable: true, get: function () { return withQueryParams_1.withQueryParamsMapped; } });
var QueryParams_1 = require("./QueryParams");
Object.defineProperty(exports, "QueryParams", { enumerable: true, get: function () { return QueryParams_1.QueryParams; } });
var QueryParamProvider_1 = require("./QueryParamProvider");
Object.defineProperty(exports, "QueryParamProvider", { enumerable: true, get: function () { return QueryParamProvider_1.QueryParamProvider; } });
