// @bundle-stats/utils/lib/webpack/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODULE_PATH_PACKAGES = exports.SUMMARY_METRIC_PATHS = exports.SECTIONS = exports.SECTION_WEBPACK_PACKAGES = exports.SECTION_WEBPACK_MODULES = exports.SECTION_WEBPACK_ASSETS = exports.SECTION_WEBPACK_SIZES = exports.SECTION_WEBPACK_STATS = void 0;
exports.SECTION_WEBPACK_STATS = 'stats';
exports.SECTION_WEBPACK_SIZES = 'sizes';
exports.SECTION_WEBPACK_ASSETS = 'assets';
exports.SECTION_WEBPACK_MODULES = 'modules';
exports.SECTION_WEBPACK_PACKAGES = 'packages';
exports.SECTIONS = [
    exports.SECTION_WEBPACK_STATS,
    exports.SECTION_WEBPACK_SIZES,
    exports.SECTION_WEBPACK_ASSETS,
    exports.SECTION_WEBPACK_MODULES,
    exports.SECTION_WEBPACK_PACKAGES,
];
exports.SUMMARY_METRIC_PATHS = [
    'totalSizeByTypeALL',
    'totalInitialSizeJS',
    'totalInitialSizeCSS',
    'cacheInvalidation',
    'moduleCount',
    'chunkCount',
    'assetCount',
    'packageCount',
    'duplicatePackagesCount',
];
exports.MODULE_PATH_PACKAGES = /(node_modules|~)\/((!?@(([\w|\-|_|.]*)\/){2})|(([\w|\-|_|.]*)\/))/g;
