// @bundle-stats/ui/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.METRICS_WEBPACK_PACKAGES = exports.METRICS_WEBPACK_MODULES = exports.METRICS_WEBPACK_ASSETS = exports.METRICS_WEBPACK_GENERAL = exports.ASSETS_SIZES_FILE_TYPE_MAP = exports.SECTION_URLS = exports.URLS = exports.SPACES = exports.SPACE_XXXLARGE = exports.SPACE_XXLARGE = exports.SPACE_XLARGE = exports.SPACE_LARGE = exports.SPACE_MEDIUM = exports.SPACE_SMALL = exports.SPACE_XSMALL = exports.SPACE_XXSMALL = exports.SPACE_XXXSMALL = exports.NO_SPACE = void 0;

var _utils = require("@bundle-stats/utils");

var _SECTION_URLS;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var NO_SPACE = 'no-space';
exports.NO_SPACE = NO_SPACE;
var SPACE_XXXSMALL = 'xxxsmall';
exports.SPACE_XXXSMALL = SPACE_XXXSMALL;
var SPACE_XXSMALL = 'xxsmall';
exports.SPACE_XXSMALL = SPACE_XXSMALL;
var SPACE_XSMALL = 'xsmall';
exports.SPACE_XSMALL = SPACE_XSMALL;
var SPACE_SMALL = 'small';
exports.SPACE_SMALL = SPACE_SMALL;
var SPACE_MEDIUM = 'medium';
exports.SPACE_MEDIUM = SPACE_MEDIUM;
var SPACE_LARGE = 'large';
exports.SPACE_LARGE = SPACE_LARGE;
var SPACE_XLARGE = 'xlarge';
exports.SPACE_XLARGE = SPACE_XLARGE;
var SPACE_XXLARGE = 'xxlarge';
exports.SPACE_XXLARGE = SPACE_XXLARGE;
var SPACE_XXXLARGE = 'xxxlarge';
exports.SPACE_XXXLARGE = SPACE_XXXLARGE;
var SPACES = [NO_SPACE, SPACE_XXXSMALL, SPACE_XXSMALL, SPACE_XSMALL, SPACE_SMALL, SPACE_MEDIUM, SPACE_LARGE, SPACE_XLARGE, SPACE_XXLARGE, SPACE_XXXLARGE];
exports.SPACES = SPACES;
var URLS = {
  OVERVIEW: '/',
  ASSETS: '/assets',
  MODULES: '/modules',
  PACKAGES: '/packages'
};
exports.URLS = URLS;
var SECTION_URLS = (_SECTION_URLS = {}, _defineProperty(_SECTION_URLS, _utils.SECTIONS.TOTALS, URLS.OVERVIEW), _defineProperty(_SECTION_URLS, _utils.SECTIONS.ASSETS, URLS.ASSETS), _defineProperty(_SECTION_URLS, _utils.SECTIONS.MODULES, URLS.MODULES), _defineProperty(_SECTION_URLS, _utils.SECTIONS.PACKAGES, URLS.PACKAGES), _SECTION_URLS);
exports.SECTION_URLS = SECTION_URLS;

var ASSETS_SIZES_FILE_TYPE_MAP = _utils.FILE_TYPES.reduce(function (agg, fileType) {
  return _objectSpread(_objectSpread({}, agg), {}, _defineProperty({}, "webpack.sizes.totalSizeByType".concat(fileType), fileType));
}, {});

exports.ASSETS_SIZES_FILE_TYPE_MAP = ASSETS_SIZES_FILE_TYPE_MAP;
var METRICS_WEBPACK_GENERAL = ['webpack.totalSizeByTypeALL', 'webpack.totalInitialSizeJS', 'webpack.totalInitialSizeCSS', 'webpack.cacheInvalidation'];
exports.METRICS_WEBPACK_GENERAL = METRICS_WEBPACK_GENERAL;
var METRICS_WEBPACK_ASSETS = ['webpack.assetCount', 'webpack.chunkCount'];
exports.METRICS_WEBPACK_ASSETS = METRICS_WEBPACK_ASSETS;
var METRICS_WEBPACK_MODULES = ['webpack.moduleCount'];
exports.METRICS_WEBPACK_MODULES = METRICS_WEBPACK_MODULES;
var METRICS_WEBPACK_PACKAGES = ['webpack.packageCount', 'webpack.duplicatePackagesCount'];
exports.METRICS_WEBPACK_PACKAGES = METRICS_WEBPACK_PACKAGES;
