// @bundle-stats/ui/lib-esm/layout/footer/footer.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import config from '../../config.json';
import { Container } from '../../ui/container';
import { Box } from '../box';
import { Stack } from '../stack';
import css from './footer.module.css';
/* eslint-disable jsx-a11y/anchor-has-content */

var ExternalLink = function ExternalLink(props) {
  return /*#__PURE__*/React.createElement("a", _extends({
    target: "_blank",
    rel: "noopener noreferrer"
  }, props));
};
/* eslint-enable jsx-a11y/anchor-has-content */


ExternalLink.propTypes = {
  /** Link to value */
  to: PropTypes.string.isRequired
};
export var Footer = function Footer(_ref) {
  var className = _ref.className,
      version = _ref.version;
  return /*#__PURE__*/React.createElement(Container, {
    as: "footer",
    className: cx(css.root, className)
  }, /*#__PURE__*/React.createElement(Box, {
    padding: ['medium', 'none'],
    className: css.inner
  }, /*#__PURE__*/React.createElement(Stack, {
    space: "xsmall"
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("nav", {
    className: css.nav
  }, /*#__PURE__*/React.createElement(ExternalLink, {
    href: config.githubUrl,
    className: cx(css.navItem, css.navLink),
    title: "In-depth bundle analyzer for webpack(bundle size, assets, modules, packages)"
  }, "BundleStats ", version), /*#__PURE__*/React.createElement("span", {
    className: css.navItem
  }, /*#__PURE__*/React.createElement(ExternalLink, {
    href: config.githubUrl + "/issues",
    className: css.navLink,
    title: "Navigate to GitHub and open issue"
  }, "Report issue")), /*#__PURE__*/React.createElement("span", {
    className: css.navItem
  }, /*#__PURE__*/React.createElement("span", null, "Made by "), /*#__PURE__*/React.createElement(ExternalLink, {
    className: css.navLink,
    href: config.siteUrl,
    title: "Go to RelativeCI - Specialized insights for web bundles"
  }, "RelativeCI")))))));
};
Footer.defaultProps = {
  className: '',
  version: null
};
Footer.propTypes = {
  /** Adopted child class name */
  className: PropTypes.string,

  /** Content */
  version: PropTypes.string
};
