// @bundle-stats/ui/lib/components/jobs-header/jobs-header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JobsHeader = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = _interopRequireDefault(require("../../i18n"));

var _flexStack = require("../../layout/flex-stack");

var _jobHeader = require("../job-header");

var _jobsHeaderModule = _interopRequireDefault(require("./jobs-header.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var JobsHeader = function JobsHeader(props) {
  var className = props.className,
      jobs = props.jobs;
  var rootClassName = (0, _classnames["default"])(_jobsHeaderModule["default"].root, className);
  return /*#__PURE__*/_react["default"].createElement(_flexStack.FlexStack, {
    className: rootClassName,
    space: "none"
  }, jobs === null || jobs === void 0 ? void 0 : jobs.map(function (job, index) {
    return /*#__PURE__*/_react["default"].createElement(_jobHeader.JobHeader, {
      key: job.internalBuildNumber || index,
      className: _jobsHeaderModule["default"].item,
      job: job,
      tag: index === 0 ? _i18n["default"].CURRENT : _i18n["default"].BASELINE
    });
  }));
};

exports.JobsHeader = JobsHeader;
JobsHeader.propTypes = {
  /** Adopted child classname */
  className: _propTypes["default"].string,

  /** Jobs data */
  jobs: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    internalBuildNumber: _propTypes["default"].number
  }))
};
JobsHeader.defaultProps = {
  className: '',
  jobs: null
};
