// @bundle-stats/ui/lib/components/job-meta/job-meta.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JobMeta = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _jobMetaModule = _interopRequireDefault(require("./job-meta.module.css"));

var _excluded = ["as", "className", "meta"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var JobMeta = function JobMeta(props) {
  var Component = props.as,
      className = props.className,
      meta = props.meta,
      restProps = _objectWithoutProperties(props, _excluded);

  var rootClassName = (0, _classnames["default"])(_jobMetaModule["default"].root, className);
  return /*#__PURE__*/_react["default"].createElement(Component, _extends({
    className: rootClassName
  }, restProps), meta.map(function (_ref, index) {
    var title = _ref.title,
        content = _ref.content;
    var key = index;
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: _jobMetaModule["default"].item,
      key: key
    }, title && /*#__PURE__*/_react["default"].createElement("h2", {
      className: _jobMetaModule["default"].itemTitle
    }, title), content.map(function (item) {
      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, item);
    }));
  }));
};

exports.JobMeta = JobMeta;
JobMeta.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string,

  /** Contrnt */
  meta: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    title: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].element]),
    content: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].element]))
  })).isRequired,

  /** Render component */
  as: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].element])
};
JobMeta.defaultProps = {
  className: '',
  as: 'div'
};
