// @bundle-stats/ui/lib-esm/components/job-meta/job-meta.js
var _excluded = ["as", "className", "meta"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import css from './job-meta.module.css';
export var JobMeta = function JobMeta(props) {
  var Component = props.as,
      className = props.className,
      meta = props.meta,
      restProps = _objectWithoutPropertiesLoose(props, _excluded);

  var rootClassName = cx(css.root, className);
  return /*#__PURE__*/React.createElement(Component, _extends({
    className: rootClassName
  }, restProps), meta.map(function (_ref, index) {
    var title = _ref.title,
        content = _ref.content;
    var key = index;
    return /*#__PURE__*/React.createElement("div", {
      className: css.item,
      key: key
    }, title && /*#__PURE__*/React.createElement("h2", {
      className: css.itemTitle
    }, title), content.map(function (item) {
      return /*#__PURE__*/React.createElement(React.Fragment, null, item);
    }));
  }));
};
JobMeta.propTypes = {
  /** Adopted child class name */
  className: PropTypes.string,

  /** Contrnt */
  meta: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    content: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.element]))
  })).isRequired,

  /** Render component */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
};
JobMeta.defaultProps = {
  className: '',
  as: 'div'
};
