// @bundle-stats/ui/lib/components/summary-item/summary-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SummaryItem = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("@bundle-stats/utils");

var _popover = require("../../ui/popover");

var _skeleton = require("../../ui/skeleton");

var _stack = require("../../layout/stack");

var _flexStack = require("../../layout/flex-stack");

var _metric = require("../metric");

var _delta = require("../delta");

var _summaryItemModule = _interopRequireDefault(require("./summary-item.module.css"));

var _excluded = ["className", "as", "size", "id", "data", "loading", "showDelta", "showMetricDescription"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var MetricInfo = function MetricInfo(_ref) {
  var description = _ref.description,
      url = _ref.url;

  // Workaround cases where the parent is a link
  var readMoreOnClick = function readMoreOnClick() {
    window.open(url);
  };

  return /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "xxxsmall"
  }, /*#__PURE__*/_react["default"].createElement("p", null, description), url && /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("button", {
    type: "button",
    className: _summaryItemModule["default"].readMoreBtn,
    onClick: readMoreOnClick
  }, "Read more")));
};

MetricInfo.propTypes = {
  description: _propTypes["default"].string.isRequired,
  url: _propTypes["default"].string
};
MetricInfo.defaultProps = {
  url: ''
};

var SummaryItem = function SummaryItem(_ref2) {
  var className = _ref2.className,
      Component = _ref2.as,
      size = _ref2.size,
      id = _ref2.id,
      data = _ref2.data,
      loading = _ref2.loading,
      showDelta = _ref2.showDelta,
      showMetricDescription = _ref2.showMetricDescription,
      props = _objectWithoutProperties(_ref2, _excluded);

  var _ref3 = data || {
    baseline: 0,
    current: 0
  },
      baseline = _ref3.baseline,
      current = _ref3.current;

  var metric = (0, _utils.getGlobalMetricType)(id);
  var runInfo = (0, _utils.getMetricRunInfo)(metric, current, baseline);
  var showMetricDescriptionTooltip = showMetricDescription && (metric === null || metric === void 0 ? void 0 : metric.description);
  var rootClassName = (0, _classnames["default"])(_summaryItemModule["default"].root, className, _summaryItemModule["default"][size], showMetricDescription && _summaryItemModule["default"].showMetricDescription, showMetricDescriptionTooltip && _summaryItemModule["default"].showMetricDescription, showDelta && _summaryItemModule["default"].showDelta);
  return /*#__PURE__*/_react["default"].createElement(_stack.Stack, _extends({
    space: "xxsmall",
    as: Component,
    className: rootClassName
  }, props), /*#__PURE__*/_react["default"].createElement(_flexStack.FlexStack, {
    as: "h3",
    space: "xxxsmall",
    className: _summaryItemModule["default"].title
  }, /*#__PURE__*/_react["default"].createElement("span", null, metric.label), showMetricDescriptionTooltip && /*#__PURE__*/_react["default"].createElement(_popover.Popover, {
    className: _summaryItemModule["default"].icon,
    icon: "help"
  }, /*#__PURE__*/_react["default"].createElement(MetricInfo, metric))), /*#__PURE__*/_react["default"].createElement(_stack.Stack, null, !loading ? /*#__PURE__*/_react["default"].createElement(_metric.Metric, {
    className: _summaryItemModule["default"].currentMetric,
    value: current,
    formatter: metric.formatter,
    enhanced: true,
    inline: true
  }, showDelta && /*#__PURE__*/_react["default"].createElement(_delta.Delta, {
    className: _summaryItemModule["default"].delta,
    displayValue: runInfo.displayDeltaPercentage,
    deltaType: runInfo.deltaType
  })) : /*#__PURE__*/_react["default"].createElement(_skeleton.Skeleton, {
    className: (0, _classnames["default"])(_summaryItemModule["default"].currentMetric, _summaryItemModule["default"].loading)
  }), !loading ? /*#__PURE__*/_react["default"].createElement(_metric.Metric, {
    className: _summaryItemModule["default"].baselineMetric,
    value: baseline,
    formatter: metric.formatter
  }) : /*#__PURE__*/_react["default"].createElement(_skeleton.Skeleton, {
    className: (0, _classnames["default"])(_summaryItemModule["default"].baselineMetric, _summaryItemModule["default"].loading)
  })));
};

exports.SummaryItem = SummaryItem;
SummaryItem.defaultProps = {
  className: '',
  as: 'div',
  data: null,
  size: 'medium',
  loading: false,
  showMetricDescription: false,
  showDelta: true
};
SummaryItem.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string,

  /** Custom component */
  as: _propTypes["default"].elementType,

  /** Size modifier */
  size: _propTypes["default"].oneOf(['medium', 'large']),

  /** Metric id */
  id: _propTypes["default"].string.isRequired,

  /** Loading flag */
  loading: _propTypes["default"].bool,

  /** Summary data */
  data: _propTypes["default"].object,
  // eslint-disable-line react/forbid-prop-types

  /** Show description */
  showMetricDescription: _propTypes["default"].bool,

  /** Show delta */
  showDelta: _propTypes["default"].bool
};
