// @bundle-stats/utils/lib/config/component-links.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPONENT = exports.SECTIONS = exports.PACKAGE_FILTERS = exports.MODULE_FILTERS = exports.ASSET_FILTERS = exports.MODULE_SOURCE_TYPE = exports.MODULE_FILE_TYPE = exports.MODULE_CHUNK = exports.ASSET_FILE_TYPE = exports.ASSET_ENTRY_TYPE = void 0;
exports.ASSET_ENTRY_TYPE = 'et';
exports.ASSET_FILE_TYPE = 'ft';
exports.MODULE_CHUNK = 'c';
exports.MODULE_FILE_TYPE = 'mft';
exports.MODULE_SOURCE_TYPE = 'mst';
exports.ASSET_FILTERS = {
    ASSET: 'asset',
    CHANGED: 'changed',
    ENTRY: 'entrypoint',
    INITIAL: 'initial',
    CHUNK: 'chunk',
};
exports.MODULE_FILTERS = {
    CHANGED: 'changed',
    FIRST_PARTY: 'fp',
    THIRD_PARTY: 'tp',
};
exports.PACKAGE_FILTERS = {
    CHANGED: 'changed',
    DUPLICATE: 'duplicate',
};
exports.SECTIONS = {
    TOTALS: 'overview',
    ASSETS: 'assets',
    MODULES: 'modules',
    PACKAGES: 'packages',
};
exports.COMPONENT = {
    BUNDLE_ASSETS: 'ba',
    BUNDLE_MODULES: 'bm',
    BUNDLE_PACKAGES: 'bp',
};
