// @bundle-stats/utils/lib/config/file-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODULE_DESTINATION_FILE_TYPES = exports.MODULE_DESTINATION_PATTERNS = exports.MODULE_SOURCE_FILE_TYPES = exports.MODULE_SOURCE_PATTERNS = exports.FILE_TYPES = exports.FILE_TYPE_LABELS = exports.FILE_TYPE_PATTERNS = exports.FILE_TYPE_OTHER = exports.FILE_TYPE_HTML = exports.FILE_TYPE_FONT = exports.FILE_TYPE_MEDIA = exports.FILE_TYPE_IMG = exports.FILE_TYPE_JS = exports.FILE_TYPE_CSS = void 0;
exports.FILE_TYPE_CSS = 'CSS';
exports.FILE_TYPE_JS = 'JS';
exports.FILE_TYPE_IMG = 'IMG';
exports.FILE_TYPE_MEDIA = 'MEDIA';
exports.FILE_TYPE_FONT = 'FONT';
exports.FILE_TYPE_HTML = 'HTML';
exports.FILE_TYPE_OTHER = 'OTHER';
exports.FILE_TYPE_PATTERNS = {
    [exports.FILE_TYPE_CSS]: /\.css$/,
    [exports.FILE_TYPE_JS]: /\.js$/,
    [exports.FILE_TYPE_IMG]: /\.(png|jpe?g|webp|gif|svg|ico)$/,
    [exports.FILE_TYPE_MEDIA]: /\.(mp4|mp3|mov)$/,
    [exports.FILE_TYPE_FONT]: /\.(woff|woff2|ttf|otf)$/,
    [exports.FILE_TYPE_HTML]: /\.(html?)$/,
};
exports.FILE_TYPE_LABELS = {
    [exports.FILE_TYPE_CSS]: 'CSS',
    [exports.FILE_TYPE_JS]: 'JS',
    [exports.FILE_TYPE_IMG]: 'IMG',
    [exports.FILE_TYPE_MEDIA]: 'Media',
    [exports.FILE_TYPE_FONT]: 'Fonts',
    [exports.FILE_TYPE_HTML]: 'HTML',
    [exports.FILE_TYPE_OTHER]: 'Other',
};
exports.FILE_TYPES = [...Object.keys(exports.FILE_TYPE_PATTERNS), exports.FILE_TYPE_OTHER];
exports.MODULE_SOURCE_PATTERNS = {
    [exports.FILE_TYPE_CSS]: /\.(css|styl|sass|scss|less)$/,
    [exports.FILE_TYPE_JS]: /\.(jsx?|tsx?|mjs)$/,
};
exports.MODULE_SOURCE_FILE_TYPES = [...Object.keys(exports.MODULE_SOURCE_PATTERNS), exports.FILE_TYPE_OTHER];
exports.MODULE_DESTINATION_PATTERNS = {
    [exports.FILE_TYPE_CSS]: /\.css$/,
    [exports.FILE_TYPE_JS]: /\.(js|mjs)$/,
};
exports.MODULE_DESTINATION_FILE_TYPES = Object.keys(exports.MODULE_DESTINATION_PATTERNS);
