// @bundle-stats/utils/lib/config/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SOURCE_PATHS = exports.SOURCE_PATH_BROWSERTIME = exports.SOURCE_PATH_LIGHTHOUSE = exports.SOURCE_PATH_WEBPACK_STATS = exports.PACKAGE_ID_SEPARATOR = exports.PACKAGES_SEPARATOR = void 0;
__exportStar(require("./component-links"), exports);
__exportStar(require("./file-types"), exports);
__exportStar(require("./insights"), exports);
__exportStar(require("./metrics"), exports);
__exportStar(require("./delta"), exports);
exports.PACKAGES_SEPARATOR = ':';
exports.PACKAGE_ID_SEPARATOR = '~';
exports.SOURCE_PATH_WEBPACK_STATS = 'webpack';
exports.SOURCE_PATH_LIGHTHOUSE = 'lighthouse';
exports.SOURCE_PATH_BROWSERTIME = 'browsertime';
exports.SOURCE_PATHS = [
    exports.SOURCE_PATH_WEBPACK_STATS,
    exports.SOURCE_PATH_LIGHTHOUSE,
    exports.SOURCE_PATH_BROWSERTIME,
];
