// @bundle-stats/utils/lib/config/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.METRIC_TYPES = exports.METRIC_TYPE_PERCENTAGE = exports.METRIC_TYPE_SCORE = exports.METRIC_TYPE_NUMERIC = exports.METRIC_TYPE_DURATION = exports.METRIC_TYPE_FILE_SIZE = void 0;
const format_1 = require("../utils/format");
exports.METRIC_TYPE_FILE_SIZE = 'METRIC_TYPE_FILE_SIZE';
exports.METRIC_TYPE_DURATION = 'METRIC_TYPE_DURATION';
exports.METRIC_TYPE_NUMERIC = 'METRIC_TYPE_NUMERIC';
exports.METRIC_TYPE_SCORE = 'METRIC_TYPE_SCORE';
exports.METRIC_TYPE_PERCENTAGE = 'METRIC_TYPE_PERCENTAGE';
exports.METRIC_TYPES = {
    [exports.METRIC_TYPE_NUMERIC]: {
        biggerIsBetter: false,
        formatter: format_1.formatNumber,
    },
    [exports.METRIC_TYPE_SCORE]: {
        biggerIsBetter: true,
        formatter: format_1.formatNumber,
    },
    [exports.METRIC_TYPE_DURATION]: {
        biggerIsBetter: false,
        formatter: format_1.formatDuration,
    },
    [exports.METRIC_TYPE_FILE_SIZE]: {
        biggerIsBetter: false,
        formatter: format_1.formatFileSize,
    },
    [exports.METRIC_TYPE_PERCENTAGE]: {
        biggerIsBetter: true,
        formatter: format_1.formatPercentage,
    },
};
