// @bundle-stats/utils/lib-esm/constants.js
export var MetricTypeType;
(function (MetricTypeType) {
    MetricTypeType[MetricTypeType["METRIC_TYPE_NUMERIC"] = 0] = "METRIC_TYPE_NUMERIC";
    MetricTypeType[MetricTypeType["METRIC_TYPE_SCORE"] = 1] = "METRIC_TYPE_SCORE";
    MetricTypeType[MetricTypeType["METRIC_TYPE_FILE_SIZE"] = 2] = "METRIC_TYPE_FILE_SIZE";
    MetricTypeType[MetricTypeType["METRIC_TYPE_DURATION"] = 3] = "METRIC_TYPE_DURATION";
    MetricTypeType[MetricTypeType["METRIC_TYPE_PERCENTAGE"] = 4] = "METRIC_TYPE_PERCENTAGE";
})(MetricTypeType || (MetricTypeType = {}));
export var MetricRunInfoDeltaType;
(function (MetricRunInfoDeltaType) {
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["HIGH_NEGATIVE"] = 0] = "HIGH_NEGATIVE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["NEGATIVE"] = 1] = "NEGATIVE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["LOW_NEGATIVE"] = 2] = "LOW_NEGATIVE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["NO_CHANGE"] = 3] = "NO_CHANGE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["LOW_POSITIVE"] = 4] = "LOW_POSITIVE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["POSITIVE"] = 5] = "POSITIVE";
    MetricRunInfoDeltaType[MetricRunInfoDeltaType["HIGH_POSITIVE"] = 6] = "HIGH_POSITIVE";
})(MetricRunInfoDeltaType || (MetricRunInfoDeltaType = {}));
