// @bundle-stats/ui/lib/components/duplicate-packages-warning/duplicate-packages-warning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DuplicatePackagesWarning = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("@bundle-stats/utils");

var _alert = require("../../ui/alert");

var _componentLink = require("../component-link");

var _delta = require("../delta");

var _duplicatePackagesWarningModule = _interopRequireDefault(require("./duplicate-packages-warning.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var numberMetric = _utils.METRIC_TYPES[_utils.METRIC_TYPE_NUMERIC];

var DuplicatePackagesWarning = function DuplicatePackagesWarning(props) {
  var className = props.className,
      duplicatePackagesCount = props.duplicatePackagesCount,
      CustomComponentLink = props.customComponentLink;
  var metricRunInfo = (0, _utils.getMetricRunInfo)(numberMetric, duplicatePackagesCount.current, duplicatePackagesCount.baseline);
  return /*#__PURE__*/_react["default"].createElement(_alert.Alert, {
    kind: "warning",
    className: (0, _classnames["default"])(_duplicatePackagesWarningModule["default"].root, className)
  }, "Bundle contains ", /*#__PURE__*/_react["default"].createElement(CustomComponentLink, _utils.BUNDLE_PACKAGES_DUPLICATE, metricRunInfo.value, metricRunInfo.delta !== 0 && /*#__PURE__*/_react["default"].createElement(_delta.Delta, {
    className: _duplicatePackagesWarningModule["default"].titleDelta,
    inverted: true,
    displayValue: metricRunInfo.displayDelta,
    deltaType: metricRunInfo.deltaType
  }), " duplicate ".concat(metricRunInfo.value === 1 ? 'package' : 'packages', ".")));
};

exports.DuplicatePackagesWarning = DuplicatePackagesWarning;
DuplicatePackagesWarning.propTypes = {
  className: _propTypes["default"].string,
  customComponentLink: _propTypes["default"].elementType,
  duplicatePackagesCount: _propTypes["default"].shape({
    current: _propTypes["default"].number,
    baseline: _propTypes["default"].number
  }).isRequired
};
DuplicatePackagesWarning.defaultProps = {
  className: '',
  customComponentLink: _componentLink.ComponentLink
};
