// @bundle-stats/ui/lib-esm/ui/empty-set/empty-set.js
import React from 'react';
import PropTypes from 'prop-types';
import I18N from '../../i18n';
import css from './empty-set.module.css';
export var EmptySet = function EmptySet(props) {
  var resources = props.resources,
      filtered = props.filtered,
      handleResetFilters = props.handleResetFilters,
      handleViewAll = props.handleViewAll;
  var message = filtered ? "No " + resources + " found." : "No " + resources + " available.";
  var actions = [].concat(handleResetFilters ? [/*#__PURE__*/React.createElement("button", {
    className: css.action,
    type: "button",
    onClick: handleResetFilters
  }, I18N.RESET_FILTERS)] : [], handleViewAll ? [/*#__PURE__*/React.createElement("button", {
    className: css.action,
    type: "button",
    onClick: handleViewAll
  }, I18N.VIEW_ALL)] : []);
  return /*#__PURE__*/React.createElement("div", {
    className: css.root
  }, /*#__PURE__*/React.createElement("p", {
    className: css.text
  }, message), filtered && actions.map(function (action, index) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, index > 0 && ' or ', action);
  }));
};
EmptySet.propTypes = {
  resources: PropTypes.string.isRequired,
  filtered: PropTypes.bool.isRequired,
  handleResetFilters: PropTypes.func,
  handleViewAll: PropTypes.func
};
EmptySet.defaultProps = {
  handleResetFilters: null,
  handleViewAll: null
};
