// @bundle-stats/ui/lib-esm/components/bundle-assets-totals-table/bundle-assets-totals-table.js
import React from 'react';
import PropTypes from 'prop-types';
import { getBundleAssetsFileTypeComponentLink } from '@bundle-stats/utils';
import * as webpack from '@bundle-stats/utils/lib-esm/webpack';
import { ASSETS_SIZES_FILE_TYPE_MAP } from '../../constants';
import { MetricsTable } from '../metrics-table';
import { ComponentLink } from '../component-link';
export var BundleAssetsTotalsTable = function BundleAssetsTotalsTable(_ref) {
  var className = _ref.className,
      jobs = _ref.jobs,
      CustomComponentLink = _ref.customComponentLink;
  var items = webpack.compareBySection.sizes(jobs);

  var renderRowHeader = function renderRowHeader(item) {
    var fileType = ASSETS_SIZES_FILE_TYPE_MAP[item.key];

    var _getBundleAssetsFileT = getBundleAssetsFileTypeComponentLink(fileType, item.label),
        section = _getBundleAssetsFileT.section,
        title = _getBundleAssetsFileT.title,
        params = _getBundleAssetsFileT.params;

    return /*#__PURE__*/React.createElement(CustomComponentLink, {
      section: section,
      title: title,
      params: params
    }, item.label);
  };

  return /*#__PURE__*/React.createElement(MetricsTable, {
    className: className,
    runs: jobs,
    items: items,
    renderRowHeader: renderRowHeader,
    showHeaderSum: true
  });
};
BundleAssetsTotalsTable.defaultProps = {
  className: '',
  jobs: [],
  customComponentLink: ComponentLink
};
BundleAssetsTotalsTable.propTypes = {
  /** Addopted child class name */
  className: PropTypes.string,

  /** Jobs data */
  jobs: PropTypes.array,
  // eslint-disable-line react/forbid-prop-types
  customComponentLink: PropTypes.elementType
};
