// @bundle-stats/ui/lib/components/total-size-type-title/total-size-type-title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TotalSizeTypeTitle = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _config = _interopRequireDefault(require("../../config.json"));

var _i18n = _interopRequireDefault(require("../../i18n"));

var _flexStack = require("../../layout/flex-stack");

var _stack = require("../../layout/stack");

var _popover = require("../../ui/popover");

var _totalSizeTypeTitleModule = _interopRequireDefault(require("./total-size-type-title.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var TotalSizeTypeTitle = function TotalSizeTypeTitle(_ref) {
  var className = _ref.className;
  return /*#__PURE__*/_react["default"].createElement(_flexStack.FlexStack, {
    space: "xxxsmall",
    className: (0, _classnames["default"])(_totalSizeTypeTitleModule["default"].title, className)
  }, /*#__PURE__*/_react["default"].createElement("span", null, _i18n["default"].ASSET_TOTALS), /*#__PURE__*/_react["default"].createElement(_popover.Popover, {
    icon: "help"
  }, /*#__PURE__*/_react["default"].createElement(_stack.Stack, {
    space: "xxxsmall"
  }, /*#__PURE__*/_react["default"].createElement("p", null, _i18n["default"].ASSET_TOTALS_INFO), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("a", {
    href: _config["default"].documentation.assets,
    target: "_blank",
    rel: "noreferrer"
  }, _i18n["default"].READ_MORE)))));
};

exports.TotalSizeTypeTitle = TotalSizeTypeTitle;
TotalSizeTypeTitle.propTypes = {
  className: _propTypes["default"].string
};
TotalSizeTypeTitle.defaultProps = {
  className: ''
};
