// @bundle-stats/ui/lib-esm/components/bundle-packages/bundle-packages.js
var _PropTypes$shape;

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { PACKAGE_FILTERS, PACKAGES_SEPARATOR, getBundleModulesBySearch, getBundlePackagesByNameComponentLink } from '@bundle-stats/utils';
import config from '../../config.json';
import I18N from '../../i18n';
import { Stack } from '../../layout/stack';
import { FlexStack } from '../../layout/flex-stack';
import { EmptySet } from '../../ui/empty-set';
import { Filters } from '../../ui/filters';
import { Popover } from '../../ui/popover';
import { SortDropdown } from '../../ui/sort-dropdown';
import { Toolbar } from '../../ui/toolbar';
import { ComponentLink } from '../component-link';
import { MetricsTable } from '../metrics-table';
import { MetricsTableSearch } from '../metrics-table-search';
import { MetricsTableOptions } from '../metrics-table-options';
import { MetricsTableTitle } from '../metrics-table-title';
import css from './bundle-packages.module.css';

var PackagePopoverContent = function PackagePopoverContent(_ref) {
  var name = _ref.name,
      fullName = _ref.fullName,
      path = _ref.path,
      duplicate = _ref.duplicate,
      CustomComponentLink = _ref.CustomComponentLink;
  var normalizedPackagePath = path || "node_modules/" + fullName.split(PACKAGES_SEPARATOR).join('/node_modules/') + "/";
  return /*#__PURE__*/React.createElement(Stack, {
    space: "xxsmall",
    className: css.packagePopover
  }, /*#__PURE__*/React.createElement("h3", {
    className: css.packagePopoverTitle
  }, name), /*#__PURE__*/React.createElement("ul", {
    className: css.packagePopoverList
  }, /*#__PURE__*/React.createElement("li", {
    className: css.packagePopoverItem
  }, /*#__PURE__*/React.createElement("a", {
    href: "https://www.npmjs.com/package/" + name,
    target: "_blank",
    rel: "noreferrer"
  }, "npmjs.com")), /*#__PURE__*/React.createElement("li", {
    className: css.packagePopoverItem
  }, /*#__PURE__*/React.createElement("a", {
    href: "https://bundlephobia.com/result?p=" + name,
    target: "_blank",
    rel: "noreferrer"
  }, "bundlephobia.com"))), /*#__PURE__*/React.createElement(Stack, {
    space: "xxxsmall",
    className: css.packagePopover.actions
  }, duplicate && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(CustomComponentLink, getBundlePackagesByNameComponentLink(name), "View all duplicate instances")), /*#__PURE__*/React.createElement(CustomComponentLink, getBundleModulesBySearch(normalizedPackagePath), "View package modules")));
};

PackagePopoverContent.propTypes = {
  name: PropTypes.string.isRequired,
  path: PropTypes.string,
  fullName: PropTypes.string.isRequired,
  duplicate: PropTypes.bool.isRequired,
  CustomComponentLink: PropTypes.elementType.isRequired
};
PackagePopoverContent.defaultProps = {
  path: ''
};

var PackageRowHeader = function PackageRowHeader(_ref2) {
  var item = _ref2.item,
      CustomComponentLink = _ref2.CustomComponentLink;
  var packageNames = item.label.split(PACKAGES_SEPARATOR);

  var _ref3 = item.runs[0] || {},
      path = _ref3.path;

  return /*#__PURE__*/React.createElement("span", {
    className: css.packageNames
  }, packageNames.map(function (packageName, index) {
    // Render duplicate flag only for the last entry
    var duplicateFlag = index === packageNames.length - 1 && item.duplicate && /*#__PURE__*/React.createElement("span", {
      className: css.duplicate,
      title: "Duplicate package"
    }, "D");
    return /*#__PURE__*/React.createElement(Popover, {
      className: css.packageName,
      icon: duplicateFlag,
      label: packageName
    }, /*#__PURE__*/React.createElement(PackagePopoverContent, {
      name: packageName,
      path: path,
      fullName: item.label,
      duplicate: item.duplicate,
      CustomComponentLink: CustomComponentLink
    }));
  }));
};

PackageRowHeader.propTypes = {
  item: PropTypes.shape({
    label: PropTypes.string,
    duplicate: PropTypes.bool,
    runs: PropTypes.arrayOf(PropTypes.shape({
      path: PropTypes.string
    })).isRequired
  }).isRequired,
  CustomComponentLink: PropTypes.elementType.isRequired
};
export var BundlePackages = function BundlePackages(props) {
  var _ref5;

  var className = props.className,
      jobs = props.jobs,
      items = props.items,
      updateFilters = props.updateFilters,
      resetFilters = props.resetFilters,
      resetAllFilters = props.resetAllFilters,
      totalRowCount = props.totalRowCount,
      filters = props.filters,
      sortItems = props.sortItems,
      sort = props.sort,
      updateSort = props.updateSort,
      search = props.search,
      updateSearch = props.updateSearch,
      hasActiveFilters = props.hasActiveFilters,
      CustomComponentLink = props.customComponentLink;
  var emptyMessage = /*#__PURE__*/React.createElement(EmptySet, {
    resources: "packages",
    filtered: totalRowCount !== 0,
    handleResetFilters: resetFilters,
    handleViewAll: resetAllFilters
  });
  var renderRowHeader = useCallback(function (item) {
    return /*#__PURE__*/React.createElement(PackageRowHeader, {
      item: item,
      CustomComponentLink: CustomComponentLink
    });
  }, [CustomComponentLink]);
  return /*#__PURE__*/React.createElement("section", {
    className: cx(css.root, className)
  }, /*#__PURE__*/React.createElement(Toolbar, {
    className: css.toolbar,
    renderActions: function renderActions(_ref4) {
      var actionClassName = _ref4.actionClassName;
      return /*#__PURE__*/React.createElement(FlexStack, {
        space: "xxsmall",
        className: cx(css.dropdown, actionClassName)
      }, /*#__PURE__*/React.createElement(SortDropdown, _extends({
        items: sortItems
      }, sort, {
        onChange: updateSort
      })), /*#__PURE__*/React.createElement(MetricsTableOptions, {
        handleViewAll: resetAllFilters,
        handleResetFilters: resetFilters
      }));
    }
  }, /*#__PURE__*/React.createElement(FlexStack, null, /*#__PURE__*/React.createElement(MetricsTableSearch, {
    className: css.toolbarSearch,
    placeholder: "Search by name",
    search: search,
    updateSearch: updateSearch
  }), /*#__PURE__*/React.createElement(Filters, {
    filters: (_ref5 = {}, _ref5[PACKAGE_FILTERS.CHANGED] = {
      label: 'Changed',
      defaultValue: filters[PACKAGE_FILTERS.CHANGED],
      disabled: jobs.length <= 1
    }, _ref5[PACKAGE_FILTERS.DUPLICATE] = {
      label: 'Duplicate',
      defaultValue: filters[PACKAGE_FILTERS.DUPLICATE]
    }, _ref5),
    label: "Filters (" + items.length + "/" + totalRowCount + ")",
    onChange: updateFilters,
    hasActiveFilters: hasActiveFilters
  }))), /*#__PURE__*/React.createElement("main", null, /*#__PURE__*/React.createElement(MetricsTable, {
    runs: jobs,
    items: items,
    emptyMessage: emptyMessage,
    renderRowHeader: renderRowHeader,
    showHeaderSum: true,
    title: /*#__PURE__*/React.createElement(MetricsTableTitle, {
      title: I18N.PACKAGES,
      info: "(" + items.length + "/" + totalRowCount + ")",
      popoverInfo: I18N.PACKAGES_INFO,
      popoverHref: config.documentation.packages
    })
  })));
};
BundlePackages.defaultProps = {
  className: '',
  totalRowCount: 0,
  hasActiveFilters: false,
  customComponentLink: ComponentLink
};
BundlePackages.propTypes = {
  className: PropTypes.string,
  jobs: PropTypes.arrayOf(PropTypes.shape({
    internalBuildNumber: PropTypes.number
  })).isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    label: PropTypes.string,
    runs: PropTypes.arrayOf(PropTypes.shape({
      displayValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      displayDelta: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    }))
  })).isRequired,
  updateFilters: PropTypes.func.isRequired,
  resetFilters: PropTypes.func.isRequired,
  resetAllFilters: PropTypes.func.isRequired,
  totalRowCount: PropTypes.number,
  filters: PropTypes.shape({
    changed: PropTypes.bool
  }).isRequired,
  hasActiveFilters: PropTypes.bool,
  sortItems: PropTypes.shape((_PropTypes$shape = {}, _PropTypes$shape[PropTypes.string] = PropTypes.shape({
    label: PropTypes.string,
    defaultDirection: PropTypes.bool
  }), _PropTypes$shape)).isRequired,
  sort: PropTypes.shape({
    sortBy: PropTypes.string,
    direction: PropTypes.string
  }).isRequired,
  updateSort: PropTypes.func.isRequired,
  search: PropTypes.string.isRequired,
  updateSearch: PropTypes.func.isRequired,
  customComponentLink: PropTypes.elementType
};
