// @bundle-stats/ui/lib/ui/table/table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Table = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _tableModule = _interopRequireDefault(require("./table.module.css"));

var _excluded = ["className"],
    _excluded2 = ["className"],
    _excluded3 = ["className"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var THead = function THead(props) {
  return /*#__PURE__*/_react["default"].createElement("thead", props);
};

var TBody = function TBody(props) {
  return /*#__PURE__*/_react["default"].createElement("tbody", props);
};

var Tr = function Tr(props) {
  var className = props.className,
      restProps = _objectWithoutProperties(props, _excluded);

  return /*#__PURE__*/_react["default"].createElement("tr", _extends({
    className: (0, _classnames["default"])(_tableModule["default"].row, className)
  }, restProps));
};

Tr.defaultProps = {
  className: ''
};
Tr.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string
};

var Th = function Th(props) {
  var className = props.className,
      restProps = _objectWithoutProperties(props, _excluded2);

  return /*#__PURE__*/_react["default"].createElement("th", _extends({
    className: (0, _classnames["default"])(_tableModule["default"].cell, className)
  }, restProps));
};

Th.defaultProps = {
  className: ''
};
Th.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string
};

var Td = function Td(props) {
  var className = props.className,
      restProps = _objectWithoutProperties(props, _excluded3);

  return /*#__PURE__*/_react["default"].createElement("td", _extends({
    className: (0, _classnames["default"])(_tableModule["default"].cell, className)
  }, restProps));
};

Td.defaultProps = {
  className: ''
};
Td.propTypes = {
  /** Adopted child class name */
  className: _propTypes["default"].string
};

var Table = function Table(_ref) {
  var className = _ref.className,
      emptyMessage = _ref.emptyMessage,
      compact = _ref.compact,
      outline = _ref.outline,
      children = _ref.children;
  var rootClassName = (0, _classnames["default"])(_tableModule["default"].root, className, compact && _tableModule["default"].compact, outline && _tableModule["default"].outline);

  if (!children) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: (0, _classnames["default"])(rootClassName, _tableModule["default"].empty)
    }, emptyMessage);
  }

  return /*#__PURE__*/_react["default"].createElement("table", {
    className: rootClassName
  }, children);
};

exports.Table = Table;
Table.THead = THead;
Table.TBody = TBody;
Table.Th = Th;
Table.Tr = Tr;
Table.Td = Td;
Table.defaultProps = {
  className: '',
  emptyMessage: 'No entries found.',
  children: null,
  compact: false,
  outline: false
};
Table.propTypes = {
  className: _propTypes["default"].string,
  emptyMessage: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].element]),
  compact: _propTypes["default"].bool,
  outline: _propTypes["default"].bool,
  children: _propTypes["default"].oneOfType([_propTypes["default"].node, _propTypes["default"].arrayOf(_propTypes["default"].node)])
};
