// @bundle-stats/ui/lib-esm/app/header/header.js
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Container } from '../../ui/container';
import { JobsHeader } from '../../components/jobs-header';
import { Logo } from '../../ui/logo';
import css from './header.module.css';
export var Header = function Header(_ref) {
  var className = _ref.className,
      jobs = _ref.jobs;
  return /*#__PURE__*/React.createElement(Container, {
    as: "header",
    className: cx(css.root, className)
  }, /*#__PURE__*/React.createElement("div", {
    className: css.inner
  }, /*#__PURE__*/React.createElement("a", {
    className: css.branding,
    href: "https://github.com/relative-ci/bundle-stats",
    target: "_blank",
    rel: "noopener noreferrer nofollow"
  }, /*#__PURE__*/React.createElement(Logo, {
    className: css.brandingLogo
  }, "BundleStats")), /*#__PURE__*/React.createElement(JobsHeader, {
    className: css.jobs,
    jobs: jobs
  })));
};
Header.propTypes = {
  className: PropTypes.string,
  jobs: PropTypes.arrayOf(PropTypes.object)
};
Header.defaultProps = {
  className: '',
  jobs: null
};
