// @bundle-stats/utils/lib/webpack/extract/assets-cache-invalidation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractAssetsCacheInvalidation = void 0;
const get_1 = __importDefault(require("lodash/get"));
const merge_1 = __importDefault(require("lodash/merge"));
const get_metric_changed_1 = require("../../report/get-metric-changed");
const merge_metrics_by_key_1 = require("../../report/merge-metrics-by-key");
const utils_1 = require("../utils");
const extractAssetsCacheInvalidation = (webpackStats, currentExtractedData, baselineBundleStats) => {
    const currentAssets = get_1.default(currentExtractedData, 'metrics.assets', {});
    const baselineAssets = get_1.default(baselineBundleStats, 'metrics.webpack.assets', {});
    const rows = merge_metrics_by_key_1.mergeMetricsByKey([currentAssets, baselineAssets]).map((row) => merge_1.default({}, row, {
        changed: get_metric_changed_1.getMetricChanged(row.runs),
        added: utils_1.getMetricAdded(row.runs),
        deleted: utils_1.getMetricDeleted(row.runs),
    }));
    const value = utils_1.calculateCacheInvalidation(rows);
    return {
        metrics: {
            cacheInvalidation: {
                value,
            },
        },
    };
};
exports.extractAssetsCacheInvalidation = extractAssetsCacheInvalidation;
