// @bundle-stats/ui/lib/app/header/header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Header = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _container = require("../../ui/container");

var _jobsHeader = require("../../components/jobs-header");

var _logo = require("../../ui/logo");

var _headerModule = _interopRequireDefault(require("./header.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Header = function Header(_ref) {
  var className = _ref.className,
      jobs = _ref.jobs;
  return /*#__PURE__*/_react["default"].createElement(_container.Container, {
    as: "header",
    className: (0, _classnames["default"])(_headerModule["default"].root, className)
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: _headerModule["default"].inner
  }, /*#__PURE__*/_react["default"].createElement("a", {
    className: _headerModule["default"].branding,
    href: "https://github.com/relative-ci/bundle-stats",
    target: "_blank",
    rel: "noopener noreferrer nofollow"
  }, /*#__PURE__*/_react["default"].createElement(_logo.Logo, {
    className: _headerModule["default"].brandingLogo
  }, "BundleStats")), /*#__PURE__*/_react["default"].createElement(_jobsHeader.JobsHeader, {
    className: _headerModule["default"].jobs,
    jobs: jobs
  })));
};

exports.Header = Header;
Header.propTypes = {
  className: _propTypes["default"].string,
  jobs: _propTypes["default"].arrayOf(_propTypes["default"].object)
};
Header.defaultProps = {
  className: '',
  jobs: null
};
